From d5ca49b57dbda0a96e5254570d782e219d4c27c6 Mon Sep 17 00:00:00 2001
From: Sven Eckelmann <sven.eckelmann@gmx.de>
Date: Mon, 24 Nov 2008 16:47:20 +0100
Subject: [PATCH] Set symbol visibility to hidden for gcc 4.0

By setting the visibility of symbols to hidden by default we can reduce the
symbol collisions. This should fix bug #260809, #297746, #299470.
---
 CMakeLists.txt                                     |    7 +
 cuneiform_src/Kern/CMakeLists.txt                  |    2 +
 cuneiform_src/Kern/ccom/CMakeLists.txt             |    1 +
 cuneiform_src/Kern/ced/CMakeLists.txt              |    1 +
 cuneiform_src/Kern/cfio/CMakeLists.txt             |    1 +
 cuneiform_src/Kern/cfio/sources/cpp/dll_main.cpp   |    1 -
 cuneiform_src/Kern/cimage/CMakeLists.txt           |    1 +
 .../Kern/cimage/sources/main/ctiimage.cpp          |    2 -
 cuneiform_src/Kern/cimage/sources/main/dll.cpp     |    2 -
 cuneiform_src/Kern/cline/CMakeLists.txt            |    1 +
 .../Kern/cline/sources/src/main/clinebase.cpp      |    1 -
 .../Kern/cline/sources/src/main/clinemain.cpp      |    1 -
 cuneiform_src/Kern/cpage/CMakeLists.txt            |    1 +
 cuneiform_src/Kern/cpage/sources/cpp/backup.h      |    1 -
 cuneiform_src/Kern/cpage/sources/cpp/compress.cpp  |    2 -
 cuneiform_src/Kern/cpage/sources/cpp/cpicture.cpp  |    2 -
 cuneiform_src/Kern/cpage/sources/cpp/data.cpp      |    2 -
 .../Kern/cpage/sources/cpp/defconvert.cpp          |    2 -
 cuneiform_src/Kern/cpage/sources/cpp/page.cpp      |    2 -
 cuneiform_src/Kern/cpage/sources/cpp/picture.cpp   |    2 -
 cuneiform_src/Kern/cpage/sources/cpp/table.cpp     |    2 -
 .../Kern/cpage/sources/cpp/tableclass.cpp          |    2 -
 cuneiform_src/Kern/cpage/sources/main/cpage.cpp    |    2 -
 cuneiform_src/Kern/cpage/sources/main/ctablex.cpp  |    2 -
 .../Kern/cpage/sources/main/dll_cpage.cpp          |    2 -
 cuneiform_src/Kern/cpu/CMakeLists.txt              |    1 +
 cuneiform_src/Kern/cpu/src/cpu32.c                 |    8 +-
 cuneiform_src/Kern/cstr/CMakeLists.txt             |    1 +
 cuneiform_src/Kern/ctb/CMakeLists.txt              |    1 +
 cuneiform_src/Kern/ctb/src/ctb_tool.c              |    4 +-
 cuneiform_src/Kern/dif/CMakeLists.txt              |    1 +
 cuneiform_src/Kern/evn32/CMakeLists.txt            |    1 +
 cuneiform_src/Kern/exc/CMakeLists.txt              |    1 +
 cuneiform_src/Kern/fon/CMakeLists.txt              |    1 +
 cuneiform_src/Kern/h/globus.h                      |   13 ++
 cuneiform_src/Kern/hdebug/__snp.c                  |   70 ++++----
 cuneiform_src/Kern/hhh/rshelllines.h               |   12 +-
 cuneiform_src/Kern/hhh/rshelllinescom.h            |   78 +++++----
 cuneiform_src/Kern/hrkint/cpu.h                    |   16 ++-
 cuneiform_src/Kern/include/winfuncs.h              |  180 ++++++++++---------
 .../Kern/itigerole/sources/tiger/h/globus.h        |   13 ++
 cuneiform_src/Kern/leo/CMakeLists.txt              |    1 +
 cuneiform_src/Kern/lns32/CMakeLists.txt            |    1 +
 cuneiform_src/Kern/loc/CMakeLists.txt              |    1 +
 cuneiform_src/Kern/loc/src/locompmn.c              |    3 +-
 cuneiform_src/Kern/mmx/CMakeLists.txt              |    1 +
 cuneiform_src/Kern/msk/CMakeLists.txt              |    1 +
 cuneiform_src/Kern/pass2/CMakeLists.txt            |    1 +
 cuneiform_src/Kern/puma/CMakeLists.txt             |    1 +
 cuneiform_src/Kern/puma/c/partlayout.cpp           |    2 -
 cuneiform_src/Kern/puma/main/dll.cpp               |    2 -
 cuneiform_src/Kern/puma/main/puma.cpp              |    1 -
 cuneiform_src/Kern/r35/CMakeLists.txt              |    1 +
 cuneiform_src/Kern/rbal/CMakeLists.txt             |    1 +
 cuneiform_src/Kern/rblock/CMakeLists.txt           |    1 +
 cuneiform_src/Kern/rblock/sources/main/_dll.cpp    |    2 -
 cuneiform_src/Kern/rblock/sources/main/_rblock.cpp |    2 -
 cuneiform_src/Kern/rcorrkegl/CMakeLists.txt        |    1 +
 cuneiform_src/Kern/rcutp/CMakeLists.txt            |    1 +
 cuneiform_src/Kern/rdib/CMakeLists.txt             |    1 +
 cuneiform_src/Kern/rdib/sources/cpp/CTDIB.cpp      |    3 -
 cuneiform_src/Kern/rfrmt/CMakeLists.txt            |    1 +
 cuneiform_src/Kern/rfrmt/sources/main/frmt.cpp     |    1 -
 cuneiform_src/Kern/rfrmt/sources/main/frmtdll.cpp  |    1 -
 cuneiform_src/Kern/rimage/CMakeLists.txt           |    1 +
 .../Kern/rimage/sources/main/criimage.cpp          |    2 -
 cuneiform_src/Kern/rimage/sources/main/dll.cpp     |    2 -
 cuneiform_src/Kern/rline/CMakeLists.txt            |    1 +
 cuneiform_src/Kern/rline/sources/dll.cpp           |    2 -
 cuneiform_src/Kern/rline/sources/findlostlines.cpp |    2 -
 cuneiform_src/Kern/rline/sources/newline.cpp       |    2 -
 cuneiform_src/Kern/rline/sources/rline.cpp         |    2 -
 cuneiform_src/Kern/rling/CMakeLists.txt            |    1 +
 cuneiform_src/Kern/rling/sources/CMakeLists.txt    |    1 +
 cuneiform_src/Kern/rling/sources/cpp/crling.cpp    |    6 -
 cuneiform_src/Kern/rling/sources/cpp/dll.cpp       |    6 -
 cuneiform_src/Kern/rmarker/CMakeLists.txt          |    1 +
 cuneiform_src/Kern/rneg/CMakeLists.txt             |    1 +
 .../rneg/sources/src/negatest/negatestcell.cpp     |    2 -
 .../Kern/rneg/sources/src/negmain/recnegmain.cpp   |    2 -
 .../Kern/rneg/sources/src/negmain/rnegbase.cpp     |    2 -
 cuneiform_src/Kern/rout/CMakeLists.txt             |    1 +
 cuneiform_src/Kern/rout/src/rout_own.h             |    2 -
 cuneiform_src/Kern/rpic/CMakeLists.txt             |    1 +
 cuneiform_src/Kern/rpic/sources/dll.cpp            |    2 -
 cuneiform_src/Kern/rpic/sources/rpic.cpp           |    2 -
 cuneiform_src/Kern/rpstr/CMakeLists.txt            |    1 +
 cuneiform_src/Kern/rreccom/CMakeLists.txt          |    1 +
 cuneiform_src/Kern/rreccom/src/cpp/recog.cpp       |    2 -
 cuneiform_src/Kern/rreccom/src/cpp/rreccom.cpp     |    2 -
 cuneiform_src/Kern/rsadd/CMakeLists.txt            |    1 +
 cuneiform_src/Kern/rselstr/CMakeLists.txt          |    1 +
 .../sources/src/chstr/addtocstr/puttocstr.cpp      |    2 -
 .../rselstr/sources/src/chstr/cont/puttocont.cpp   |    3 -
 .../sources/src/chstr/cutstr/chstr_cutstr.cpp      |    2 -
 .../rselstr/sources/src/chstr/rotate/rotate.cpp    |    1 -
 .../sources/src/chstr/vertical/testforvertical.cpp |    2 -
 .../Kern/rselstr/sources/src/cpp/cutstr.cpp        |    1 -
 .../sources/src/rselstrmain/rselstrbase.cpp        |    1 -
 .../sources/src/rselstrmain/rselstrmain.cpp        |    3 -
 cuneiform_src/Kern/rshelllines/CMakeLists.txt      |    1 +
 cuneiform_src/Kern/rshelllines/src/rshelllines.cpp |   36 ++--
 cuneiform_src/Kern/rstr/CMakeLists.txt             |    1 +
 cuneiform_src/Kern/rstuff/CMakeLists.txt           |    1 +
 cuneiform_src/Kern/rstuff/sources/main/dll.cpp     |    2 -
 cuneiform_src/Kern/rverline/CMakeLists.txt         |    1 +
 cuneiform_src/Kern/rverline/src/root/vl_kern.cpp   |    2 -
 cuneiform_src/Kern/rverline/src/root/vl_resid.cpp  |    2 -
 cuneiform_src/Kern/smetric/CMakeLists.txt          |    1 +
 cuneiform_src/Kern/std/CMakeLists.txt              |    1 +
 cuneiform_src/Kern/windummy.c                      |    2 +-
 111 files changed, 301 insertions(+), 293 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 63cf265..04f973b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -26,5 +26,12 @@ else()
     set(LIBDIR "lib")
 endif()
 
+# Hide non exported functions/variables
+include(CheckCCompilerFlag)
+check_c_compiler_flag("-fvisibility=hidden -DHAVE_GCCVISIBILITY" HAVE_GCCVISIBILITY)
+if (HAVE_GCCVISIBILITY)
+	add_definitions("-fvisibility=hidden -DHAVE_GCCVISIBILITY")
+endif (HAVE_GCCVISIBILITY)
+
 add_subdirectory(datafiles)
 add_subdirectory(cuneiform_src)
diff --git a/cuneiform_src/Kern/CMakeLists.txt b/cuneiform_src/Kern/CMakeLists.txt
index 1c160cc..beaf8e6 100644
--- a/cuneiform_src/Kern/CMakeLists.txt
+++ b/cuneiform_src/Kern/CMakeLists.txt
@@ -1,3 +1,4 @@
+add_definitions("-D__DPUMA__")
 include_directories(include
 h
 hh
@@ -107,6 +108,7 @@ cimage loc rdib cfio)
 #target_link_libraries(rpstrtest ${RPSTR_LIBS})
 
 add_library(windummy ${LIBTYPE} windummy.c)
+set_property(TARGET windummy PROPERTY COMPILE_DEFINITIONS __WINDUMMY__)
 target_link_libraries(windummy ${DL_LIB})
 
 add_executable(cuneiform cuneiform-cli.cpp)
diff --git a/cuneiform_src/Kern/ccom/CMakeLists.txt b/cuneiform_src/Kern/ccom/CMakeLists.txt
index 66a8ba5..902ae1b 100644
--- a/cuneiform_src/Kern/ccom/CMakeLists.txt
+++ b/cuneiform_src/Kern/ccom/CMakeLists.txt
@@ -3,6 +3,7 @@ include_directories(src)
 add_library(ccom ${LIBTYPE}
 src/ccom.c
 src/ccom_loc.c)
+set_property(TARGET ccom PROPERTY COMPILE_DEFINITIONS __CCOM__)
 
 if(WIN32)
   install(TARGETS ccom ARCHIVE DESTINATION ${LIBDIR})
diff --git a/cuneiform_src/Kern/ced/CMakeLists.txt b/cuneiform_src/Kern/ced/CMakeLists.txt
index f8de1fb..8c2106d 100644
--- a/cuneiform_src/Kern/ced/CMakeLists.txt
+++ b/cuneiform_src/Kern/ced/CMakeLists.txt
@@ -10,6 +10,7 @@ sources/main/dll.cpp
 sources/main/stdafx.cpp
 sources/main/wrapper.cpp
 )
+set_property(TARGET ced PROPERTY COMPILE_DEFINITIONS __CED__)
 
 target_link_libraries(ced cfio windummy)
 
diff --git a/cuneiform_src/Kern/cfio/CMakeLists.txt b/cuneiform_src/Kern/cfio/CMakeLists.txt
index 0abbdb2..fd58436 100644
--- a/cuneiform_src/Kern/cfio/CMakeLists.txt
+++ b/cuneiform_src/Kern/cfio/CMakeLists.txt
@@ -8,6 +8,7 @@ sources/cpp/ctcmemory.cpp
 sources/cpp/ctcstorage.cpp
 sources/cpp/dll_main.cpp
 )
+set_property(TARGET cfio PROPERTY COMPILE_DEFINITIONS __CFIO__)
 
 target_link_libraries(cfio windummy)
 
diff --git a/cuneiform_src/Kern/cfio/sources/cpp/dll_main.cpp b/cuneiform_src/Kern/cfio/sources/cpp/dll_main.cpp
index 04020c5..c6e33e4 100644
--- a/cuneiform_src/Kern/cfio/sources/cpp/dll_main.cpp
+++ b/cuneiform_src/Kern/cfio/sources/cpp/dll_main.cpp
@@ -67,7 +67,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //                    started at 25 may 1998                                    //
 //                                                                              //
 //////////////////////////////////////////////////////////////////////////////////
-#define __CFIO__
 //#define CFIO_USE_WIN32_API
 #include "ctccontrol.h"
 //#undef CFIO_USE_WIN32_API
diff --git a/cuneiform_src/Kern/cimage/CMakeLists.txt b/cuneiform_src/Kern/cimage/CMakeLists.txt
index e59d519..a5cb591 100644
--- a/cuneiform_src/Kern/cimage/CMakeLists.txt
+++ b/cuneiform_src/Kern/cimage/CMakeLists.txt
@@ -12,6 +12,7 @@ sources/main/ctimasklinesegment.cpp
 sources/main/ctimemory.cpp
 sources/main/dll.cpp
 )
+set_property(TARGET cimage PROPERTY COMPILE_DEFINITIONS __CIMAGE__)
 
 target_link_libraries(cimage cfio rdib)
 
diff --git a/cuneiform_src/Kern/cimage/sources/main/ctiimage.cpp b/cuneiform_src/Kern/cimage/sources/main/ctiimage.cpp
index ff1fb0a..789ec0e 100644
--- a/cuneiform_src/Kern/cimage/sources/main/ctiimage.cpp
+++ b/cuneiform_src/Kern/cimage/sources/main/ctiimage.cpp
@@ -54,8 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __CIMAGE__
-
 #include "resource.h"
 #include "ctidefines.h"
 #include "ctiimage.h"
diff --git a/cuneiform_src/Kern/cimage/sources/main/dll.cpp b/cuneiform_src/Kern/cimage/sources/main/dll.cpp
index d48f5aa..7331c32 100644
--- a/cuneiform_src/Kern/cimage/sources/main/dll.cpp
+++ b/cuneiform_src/Kern/cimage/sources/main/dll.cpp
@@ -61,8 +61,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //
 // dll_cpage.cpp :
 // ============================================================================
-#define __CIMAGE__
-
 /*#include <windows.h>*/
 
 #include "resource.h"
diff --git a/cuneiform_src/Kern/cline/CMakeLists.txt b/cuneiform_src/Kern/cline/CMakeLists.txt
index 12a40f8..1f1d3b1 100644
--- a/cuneiform_src/Kern/cline/CMakeLists.txt
+++ b/cuneiform_src/Kern/cline/CMakeLists.txt
@@ -8,6 +8,7 @@ sources/src/main/clinemain.cpp
 sources/src/cpp/debug.cpp
 sources/src/cpp/lcline.cpp
 )
+set_property(TARGET cline PROPERTY COMPILE_DEFINITIONS __CLINE__)
 
 target_link_libraries(cline windummy)
 
diff --git a/cuneiform_src/Kern/cline/sources/src/main/clinebase.cpp b/cuneiform_src/Kern/cline/sources/src/main/clinebase.cpp
index eb4be37..b7f5332 100644
--- a/cuneiform_src/Kern/cline/sources/src/main/clinebase.cpp
+++ b/cuneiform_src/Kern/cline/sources/src/main/clinebase.cpp
@@ -54,7 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __CLINE__
 /*#include <crtdbg.h>*/
 #include <assert.h>
 #include <stdlib.h>
diff --git a/cuneiform_src/Kern/cline/sources/src/main/clinemain.cpp b/cuneiform_src/Kern/cline/sources/src/main/clinemain.cpp
index f19471f..f62543d 100644
--- a/cuneiform_src/Kern/cline/sources/src/main/clinemain.cpp
+++ b/cuneiform_src/Kern/cline/sources/src/main/clinemain.cpp
@@ -55,7 +55,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
 /*#include<windows.h>*/
-#define __CLINE__
 
 #include<stdio.h>
 #include "clinefunc.h"
diff --git a/cuneiform_src/Kern/cpage/CMakeLists.txt b/cuneiform_src/Kern/cpage/CMakeLists.txt
index 72319b6..e790915 100644
--- a/cuneiform_src/Kern/cpage/CMakeLists.txt
+++ b/cuneiform_src/Kern/cpage/CMakeLists.txt
@@ -18,6 +18,7 @@ sources/cpp/picture.cpp
 sources/cpp/table.cpp
 sources/cpp/tableclass.cpp
 )
+set_property(TARGET cpage PROPERTY COMPILE_DEFINITIONS __CPAGE__)
 
 target_link_libraries(cpage windummy)
 
diff --git a/cuneiform_src/Kern/cpage/sources/cpp/backup.h b/cuneiform_src/Kern/cpage/sources/cpp/backup.h
index 88e142b..af3d3cd 100644
--- a/cuneiform_src/Kern/cpage/sources/cpp/backup.h
+++ b/cuneiform_src/Kern/cpage/sources/cpp/backup.h
@@ -57,7 +57,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #ifndef __BACKUP_H__
 #define __BACKUP_H__
 
-#define __CPAGE__
 //#define DPUMA_ON 1
 
 #include "page.h"
diff --git a/cuneiform_src/Kern/cpage/sources/cpp/compress.cpp b/cuneiform_src/Kern/cpage/sources/cpp/compress.cpp
index b9776da..272efeb 100644
--- a/cuneiform_src/Kern/cpage/sources/cpp/compress.cpp
+++ b/cuneiform_src/Kern/cpage/sources/cpp/compress.cpp
@@ -56,8 +56,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 /*#include <windows.h>*/
 
-#define __CPAGE__
-
 #include "globus.h"
 #include "cpage.h"
 #include "backup.h"
diff --git a/cuneiform_src/Kern/cpage/sources/cpp/cpicture.cpp b/cuneiform_src/Kern/cpage/sources/cpp/cpicture.cpp
index c06982d..79641a7 100644
--- a/cuneiform_src/Kern/cpage/sources/cpp/cpicture.cpp
+++ b/cuneiform_src/Kern/cpage/sources/cpp/cpicture.cpp
@@ -54,8 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __CPAGE__
-
 /*#include <windows.h>*/
 #include <stdlib.h>
 
diff --git a/cuneiform_src/Kern/cpage/sources/cpp/data.cpp b/cuneiform_src/Kern/cpage/sources/cpp/data.cpp
index 40f4978..2952dbb 100644
--- a/cuneiform_src/Kern/cpage/sources/cpp/data.cpp
+++ b/cuneiform_src/Kern/cpage/sources/cpp/data.cpp
@@ -58,8 +58,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <memory.h>
 #include <string.h>
 
-#define __CPAGE__
-
 #include "cpage.h"
 #include "data.h"
 #include "mymem.h"
diff --git a/cuneiform_src/Kern/cpage/sources/cpp/defconvert.cpp b/cuneiform_src/Kern/cpage/sources/cpp/defconvert.cpp
index e468ba3..ac8cca1 100644
--- a/cuneiform_src/Kern/cpage/sources/cpp/defconvert.cpp
+++ b/cuneiform_src/Kern/cpage/sources/cpp/defconvert.cpp
@@ -56,8 +56,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 /*#include <windows.h>*/
 
-#define __CPAGE__
-
 #include "globus.h"
 #include "cpage.h"
 #include "backup.h"
diff --git a/cuneiform_src/Kern/cpage/sources/cpp/page.cpp b/cuneiform_src/Kern/cpage/sources/cpp/page.cpp
index 13d0cbb..5a77cd9 100644
--- a/cuneiform_src/Kern/cpage/sources/cpp/page.cpp
+++ b/cuneiform_src/Kern/cpage/sources/cpp/page.cpp
@@ -56,8 +56,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #include <memory.h>
 
-#define __CPAGE__
-
 #include "backup.h"
 #include "page.h"
 
diff --git a/cuneiform_src/Kern/cpage/sources/cpp/picture.cpp b/cuneiform_src/Kern/cpage/sources/cpp/picture.cpp
index fab6d21..8f80aa4 100644
--- a/cuneiform_src/Kern/cpage/sources/cpp/picture.cpp
+++ b/cuneiform_src/Kern/cpage/sources/cpp/picture.cpp
@@ -56,8 +56,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 /*#include <windows.h>*/
 
-#define __CPAGE__
-
 #include "globus.h"
 #include "cpage.h"
 #include "backup.h"
diff --git a/cuneiform_src/Kern/cpage/sources/cpp/table.cpp b/cuneiform_src/Kern/cpage/sources/cpp/table.cpp
index 80c7d36..37c761e 100644
--- a/cuneiform_src/Kern/cpage/sources/cpp/table.cpp
+++ b/cuneiform_src/Kern/cpage/sources/cpp/table.cpp
@@ -56,8 +56,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 /*#include <windows.h>*/
 
-#define __CPAGE__
-
 #include "globus.h"
 #include "cpage.h"
 #include "backup.h"
diff --git a/cuneiform_src/Kern/cpage/sources/cpp/tableclass.cpp b/cuneiform_src/Kern/cpage/sources/cpp/tableclass.cpp
index 4c936c1..d1e6c28 100644
--- a/cuneiform_src/Kern/cpage/sources/cpp/tableclass.cpp
+++ b/cuneiform_src/Kern/cpage/sources/cpp/tableclass.cpp
@@ -57,8 +57,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 // TableClass.cpp: implementation of the TableClass class.
 //
 //////////////////////////////////////////////////////////////////////
-#define __CPAGE__
-
 #include "mymem.h"
 #include "tableclass.h"
 //////////////////////////////////////////////////////////////////////
diff --git a/cuneiform_src/Kern/cpage/sources/main/cpage.cpp b/cuneiform_src/Kern/cpage/sources/main/cpage.cpp
index 249bf4b..7d6be8d 100644
--- a/cuneiform_src/Kern/cpage/sources/main/cpage.cpp
+++ b/cuneiform_src/Kern/cpage/sources/main/cpage.cpp
@@ -54,8 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __CPAGE__
-
 #include "resource.h"
 #include "cpage.h"
 #include "backup.h"
diff --git a/cuneiform_src/Kern/cpage/sources/main/ctablex.cpp b/cuneiform_src/Kern/cpage/sources/main/ctablex.cpp
index ba8316d..a34d762 100644
--- a/cuneiform_src/Kern/cpage/sources/main/ctablex.cpp
+++ b/cuneiform_src/Kern/cpage/sources/main/ctablex.cpp
@@ -56,8 +56,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 /*#include <windows.h>*/
 
-#define __CPAGE__
-
 #include "resource.h"
 #include "mymem.h"
 #include "cpage.h"
diff --git a/cuneiform_src/Kern/cpage/sources/main/dll_cpage.cpp b/cuneiform_src/Kern/cpage/sources/main/dll_cpage.cpp
index e281354..218d4f2 100644
--- a/cuneiform_src/Kern/cpage/sources/main/dll_cpage.cpp
+++ b/cuneiform_src/Kern/cpage/sources/main/dll_cpage.cpp
@@ -64,8 +64,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 /*#include <windows.h>*/
 
-#define __CPAGE__
-
 #include "resource.h"
 #include "cpage.h"
 #include "backup.h"
diff --git a/cuneiform_src/Kern/cpu/CMakeLists.txt b/cuneiform_src/Kern/cpu/CMakeLists.txt
index 5e7dad5..a15781d 100644
--- a/cuneiform_src/Kern/cpu/CMakeLists.txt
+++ b/cuneiform_src/Kern/cpu/CMakeLists.txt
@@ -2,6 +2,7 @@ add_library(cpu ${LIBTYPE}
 src/cpu.c
 src/cpu32.c
 )
+set_property(TARGET cpu PROPERTY COMPILE_DEFINITIONS __CPU__)
 
 if(WIN32)
   install(TARGETS cpu ARCHIVE DESTINATION ${LIBDIR})
diff --git a/cuneiform_src/Kern/cpu/src/cpu32.c b/cuneiform_src/Kern/cpu/src/cpu32.c
index f5b81a3..4e7d5c4 100644
--- a/cuneiform_src/Kern/cpu/src/cpu32.c
+++ b/cuneiform_src/Kern/cpu/src/cpu32.c
@@ -54,11 +54,11 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-//#include "cpu.h"
+#include "cpu.h"
 #include "decl.h"
 #define PC_TYPE
 
-__declspec( dllexport ) int GetCPUName(void)
+CPU_FUNC(int) GetCPUName(void)
 {
 #ifdef PC_TYPE
 
@@ -69,7 +69,7 @@ __declspec( dllexport ) int GetCPUName(void)
 #endif
 }
 
-__declspec( dllexport ) int Get_MMX(void)
+CPU_FUNC(int) Get_MMX(void)
 {
 #ifdef PC_TYPE
 
@@ -80,7 +80,7 @@ __declspec( dllexport ) int Get_MMX(void)
 #endif
 }
 
-__declspec( dllexport ) int Get_XMM(void)
+CPU_FUNC(int) Get_XMM(void)
 {
 #ifdef PC_TYPE
 
diff --git a/cuneiform_src/Kern/cstr/CMakeLists.txt b/cuneiform_src/Kern/cstr/CMakeLists.txt
index b542b36..ef4ec10 100644
--- a/cuneiform_src/Kern/cstr/CMakeLists.txt
+++ b/cuneiform_src/Kern/cstr/CMakeLists.txt
@@ -3,6 +3,7 @@ include_directories(src)
 add_library(cstr ${LIBTYPE}
 src/cgraph.cpp
 src/cstr.c)
+set_property(TARGET cstr PROPERTY COMPILE_DEFINITIONS __CSTR__)
 
 target_link_libraries(cstr ccom)
 
diff --git a/cuneiform_src/Kern/ctb/CMakeLists.txt b/cuneiform_src/Kern/ctb/CMakeLists.txt
index 6bfefca..4070085 100644
--- a/cuneiform_src/Kern/ctb/CMakeLists.txt
+++ b/cuneiform_src/Kern/ctb/CMakeLists.txt
@@ -8,6 +8,7 @@ src/ctb_prot.c
 src/ctb_recr.c
 src/ctb_tool.c
 )
+set_property(TARGET ctb32 PROPERTY COMPILE_DEFINITIONS __CTB__)
 
 target_link_libraries(ctb32 windummy)
 
diff --git a/cuneiform_src/Kern/ctb/src/ctb_tool.c b/cuneiform_src/Kern/ctb/src/ctb_tool.c
index aad7710..b4f5dce 100644
--- a/cuneiform_src/Kern/ctb/src/ctb_tool.c
+++ b/cuneiform_src/Kern/ctb/src/ctb_tool.c
@@ -116,8 +116,8 @@ Bool32  CTB_files_init(char *file_name,Word8 *data,Int16 maxX,Int16 maxY,
 //********************************************************************
 Int32   ctb_err_code = CTB_ERR_NONE;   // error code
 char * ctb_tmp_dir=NULL;
-char local_grey_ctb[256] = "page6666";
-char local_ctb_name[256] = "ct666666";
+CTB_FUNC(char) local_grey_ctb[256] = "page6666";
+CTB_FUNC(char) local_ctb_name[256] = "ct666666";
 //********************************************************************
 //***********  EXPORT functions from CTB_pack ************************
 //********************************************************************
diff --git a/cuneiform_src/Kern/dif/CMakeLists.txt b/cuneiform_src/Kern/dif/CMakeLists.txt
index c83e8c1..996f4ea 100644
--- a/cuneiform_src/Kern/dif/CMakeLists.txt
+++ b/cuneiform_src/Kern/dif/CMakeLists.txt
@@ -8,6 +8,7 @@ src/sticchar.c
 src/sticdif.c
 src/stictool.c
 )
+set_property(TARGET dif PROPERTY COMPILE_DEFINITIONS __DIF__)
 
 if(WIN32)
   install(TARGETS dif ARCHIVE DESTINATION ${LIBDIR})
diff --git a/cuneiform_src/Kern/evn32/CMakeLists.txt b/cuneiform_src/Kern/evn32/CMakeLists.txt
index 51fd924..e2b8a7c 100644
--- a/cuneiform_src/Kern/evn32/CMakeLists.txt
+++ b/cuneiform_src/Kern/evn32/CMakeLists.txt
@@ -12,6 +12,7 @@ src/v0compev.c
 src/v0compgl.c
 src/v0comprq.c
 )
+set_property(TARGET evn32 PROPERTY COMPILE_DEFINITIONS __EVN__)
 
 target_link_libraries(evn32 dif windummy)
 
diff --git a/cuneiform_src/Kern/exc/CMakeLists.txt b/cuneiform_src/Kern/exc/CMakeLists.txt
index 5ada683..e0350fa 100644
--- a/cuneiform_src/Kern/exc/CMakeLists.txt
+++ b/cuneiform_src/Kern/exc/CMakeLists.txt
@@ -10,6 +10,7 @@ src/v0compan.c
 src/v0compgl.c
 src/v0compmn.c
 )
+set_property(TARGET exc PROPERTY COMPILE_DEFINITIONS __EXC__)
 
 target_link_libraries(exc loc ccom evn32)
 
diff --git a/cuneiform_src/Kern/fon/CMakeLists.txt b/cuneiform_src/Kern/fon/CMakeLists.txt
index 859c11c..46d8381 100644
--- a/cuneiform_src/Kern/fon/CMakeLists.txt
+++ b/cuneiform_src/Kern/fon/CMakeLists.txt
@@ -15,6 +15,7 @@ src/p2_mem.c
 src/p2_thick.c
 src/utilctb.c
 )
+set_property(TARGET fon PROPERTY COMPILE_DEFINITIONS __FON__)
 
 target_link_libraries(fon cstr ctb32)
 
diff --git a/cuneiform_src/Kern/h/globus.h b/cuneiform_src/Kern/h/globus.h
index 910698f..5a2050e 100644
--- a/cuneiform_src/Kern/h/globus.h
+++ b/cuneiform_src/Kern/h/globus.h
@@ -152,6 +152,19 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
             #define CDECL      __cdecl
          #endif
       #endif
+   #elif    defined( __GNUC_MINOR__ ) && defined(HAVE_GCCVISIBILITY) /* GNU Compiler Suite ******/
+         #define CLA_IMPO
+         #define CLA_EXPO __attribute__ ((visibility("default")))
+         #define FUN_IMPO__
+         #define __FUN_IMPO
+         #define FUN_EXPO__ __attribute__ ((visibility("default")))
+         #define __FUN_EXPO
+         #ifndef PASCAL
+            #define PASCAL __attribute__ ((stdcall))
+         #endif
+         #ifndef CDECL
+            #define CDECL __attribute__ ((cdecl))
+         #endif
    #else /* unknown compiler ******************************************/
 	     #define CLA_IMPO
          #define CLA_EXPO
diff --git a/cuneiform_src/Kern/hdebug/__snp.c b/cuneiform_src/Kern/hdebug/__snp.c
index ecb6b56..ee5cbf4 100644
--- a/cuneiform_src/Kern/hdebug/__snp.c
+++ b/cuneiform_src/Kern/hdebug/__snp.c
@@ -179,7 +179,7 @@ static int __DPUMA__AllocHook__( int allocType, void *userData, size_t size, int
 }
 
 //////////////////////////////////////////////
-Bool32   LDPUMA_Init(Word16 wHightCode, Handle hStorage)
+DPUMA_FUNC(Bool32)   LDPUMA_Init(Word16 wHightCode, Handle hStorage)
 {
 Bool32 rc = FALSE;
 #ifdef _DEBUG
@@ -290,7 +290,7 @@ Bool32 rc = FALSE;
 	return rc;
 }
 //////////////////////////////////////////////
-Bool32 LDPUMA_Done()
+DPUMA_FUNC(Bool32) LDPUMA_Done()
 {
 	Bool32 rc = FALSE;
 #ifdef _DEBUG
@@ -346,25 +346,25 @@ Word32   LDPUMA_CreateSnap()
 	return rc;
 }
 //////////////////////////////////////////////
-void LDPUMA_Stop()
+DPUMA_FUNC(void) LDPUMA_Stop()
 {
 	if(Stop)
 		 Stop();
 }
 //////////////////////////////////////////////
-Handle LDPUMA_CreateWindow(const char * lpName, void * lpDIB)
+DPUMA_FUNC(Handle) LDPUMA_CreateWindow(const char * lpName, void * lpDIB)
 {
 	if(fCreateWindow)
 		return fCreateWindow(lpName,lpDIB);
 	return NULL;
 }
 //////////////////////////////////////////////
-Bool32 LDPUMA_IsActive(void)
+DPUMA_FUNC(Bool32) LDPUMA_IsActive(void)
 {
 	return IsActive ? IsActive():FALSE;
 };
 //////////////////////////////////////////////
-void   LDPUMA_DrawRect(Handle wnd,Rect16* rc, Int32 skew, Word32 rgb_color,
+DPUMA_FUNC(void)   LDPUMA_DrawRect(Handle wnd,Rect16* rc, Int32 skew, Word32 rgb_color,
       Int16 pen_width,Word32 key)
 {
 	if(DrawRect)
@@ -378,13 +378,13 @@ void   LDPUMA_DrawRectTip(Handle wnd,Rect16* rc, Int32 skew, Word32 rgb_color,
 		DrawRectTip(wnd,rc,skew,rgb_color,(Int16)pen_width,key,pTip);
 }
 //////////////////////////////////////////////
-void   LDPUMA_DeleteRects(Handle wnd, Word32 key)
+DPUMA_FUNC(void)   LDPUMA_DeleteRects(Handle wnd, Word32 key)
 {
 	if(DeleteRects)
 		DeleteRects(wnd,key);
 };
 //////////////////////////////////////////////
-void  LDPUMA_DrawLine(Handle wnd,Point16* start, Point16* end, Int32 skew,
+DPUMA_FUNC(void)  LDPUMA_DrawLine(Handle wnd,Point16* start, Point16* end, Int32 skew,
       Word32 rgb_color,Int16 pen_width,Word32 key )
 {
 	if(DrawLine)
@@ -398,7 +398,7 @@ void  LDPUMA_DrawLineTip(Handle wnd,Point16* start, Point16* end, Int32 skew,
 		DrawLineTip(wnd,start,end,skew,rgb_color,pen_width,key,pTip);
 };
 //////////////////////////////////////////////
-void   LDPUMA_DeleteLines(Handle wnd,Word32 key)
+DPUMA_FUNC(void)   LDPUMA_DeleteLines(Handle wnd,Word32 key)
 {
 	if(DeleteLines)
 		DeleteLines(wnd,key);
@@ -410,7 +410,7 @@ void   LDPUMA_UpdateView(Handle wnd)
 		UpdateView(wnd);
 }
 //////////////////////////////////////////////
-Int32 LDPUMA_Console(const char * message,...)
+DPUMA_FUNC(Int32) LDPUMA_Console(const char * message,...)
 {
     va_list marker;
     va_start(marker, message);
@@ -430,7 +430,7 @@ Int32 LDPUMA_Console(const char * message,...)
 	*/
 
 }
-Int32 LDPUMA_ConsoleN(const char * message,...)
+DPUMA_FUNC(Int32) LDPUMA_ConsoleN(const char * message,...)
 {
 	Int32 rc = 0;
 	if(Console)
@@ -495,7 +495,7 @@ Bool16 LDPUMA_GetUserPoint(Handle wnd,Point16* pnt)
 	return rc;
 };
 //////////////////////////////////////////////
-Word32 LDPUMA_WaitUserInput(Handle cur_node,Handle wnd)
+DPUMA_FUNC(Word32) LDPUMA_WaitUserInput(Handle cur_node,Handle wnd)
 {
 	Word32 rc = 0;
 	if(WaitUserInput)
@@ -503,25 +503,25 @@ Word32 LDPUMA_WaitUserInput(Handle cur_node,Handle wnd)
 	return rc;
 };
 //////////////////////////////////////////////
-void LDPUMA_Registry( Handle node, const char * name,Handle parent )
+DPUMA_FUNC(void) LDPUMA_Registry( Handle node, const char * name,Handle parent )
 {
 	if(Registry)
 		Registry(node,name,parent);
 };
 //////////////////////////////////////////////
-void LDPUMA_StartLoop( Handle node, Word32 iter_total )
+DPUMA_FUNC(void) LDPUMA_StartLoop( Handle node, Word32 iter_total )
 {
 	if(StartLoop)
 		StartLoop(node,iter_total);
 };
 //////////////////////////////////////////////
-void LDPUMA_LoopNext( Handle node )
+DPUMA_FUNC(void) LDPUMA_LoopNext( Handle node )
 {
 	if(LoopNext)
 		LoopNext(node);
 };
 //////////////////////////////////////////////
-Bool16  LDPUMA_Skip( Handle node )
+DPUMA_FUNC(Bool16)  LDPUMA_Skip( Handle node )
 {
 	return Skip ? Skip(node):TRUE;
 };
@@ -532,7 +532,7 @@ void LDPUMA_DrawRaster(DPUMA_RecRaster * raster)
 		DrawRaster((DPUMA_RecRaster *)raster);
 };
 //////////////////////////////////////////////
-void LDPUMA_RasterText(char * lpText)
+DPUMA_FUNC(void) LDPUMA_RasterText(char * lpText)
 {
 	if(RasterText)
 		RasterText(lpText);
@@ -556,7 +556,7 @@ void   LDPUMA_DrawFocusRect(Handle wnd,Rect16* rc)
 		fnDrawFocusRect(wnd,rc);
 };
 //////////////////////////////////////////////
-Bool32 LDPUMA_RegVariable(Handle owner, const char * lpText,void * lpData, const char * lpType)
+DPUMA_FUNC(Bool32) LDPUMA_RegVariable(Handle owner, const char * lpText,void * lpData, const char * lpType)
 {
 	BOOL rc = FALSE;
 	if(RegVariable)
@@ -564,7 +564,7 @@ Bool32 LDPUMA_RegVariable(Handle owner, const char * lpText,void * lpData, const
 	return rc;
 };
 
-void LDPUMA_UnregVariable(void * lpData)
+DPUMA_FUNC(void) LDPUMA_UnregVariable(void * lpData)
 {
 	if(UnregVariable)
 		UnregVariable(lpData);
@@ -616,7 +616,7 @@ DPUMA_Callback_WindowProc   LDPUMA_SetCallbackMainFrameWindowProc(DPUMA_Callback
 	return NULL;
 };
 //////////////////////////////////////////////
-void  LDPUMA_DestroyWindow(Handle wnd)
+DPUMA_FUNC(void)  LDPUMA_DestroyWindow(Handle wnd)
 {
 	if(fDestroyWindow)
 		fDestroyWindow( wnd );
@@ -638,7 +638,7 @@ Word32  LDPUMA_SendMainWnd(Word32 message, Word32 wParam, Word32 lParam)
 	return rc;
 };
 //////////////////////////////////////////////
-Word32 LDPUMA_CSTR_Monitor(Handle owner, Word32 cstr_line, Word32 pos,
+DPUMA_FUNC(Word32) LDPUMA_CSTR_Monitor(Handle owner, Word32 cstr_line, Word32 pos,
 									  DPUMA_Callback_WindowProc lpproc)
 {
 	Word32 rc = 0;
@@ -647,7 +647,7 @@ Word32 LDPUMA_CSTR_Monitor(Handle owner, Word32 cstr_line, Word32 pos,
 return rc;
 }
 //////////////////////////////////////////////
-Word32 LDPUMA_CSTR_GetPosition(Word32 * cstr_raster)
+DPUMA_FUNC(Word32) LDPUMA_CSTR_GetPosition(Word32 * cstr_raster)
 {
 	Word32 rc = 0;
 	if(cstr_GetPosition)
@@ -662,7 +662,7 @@ void LDPUMA_CSTR_SetPosition(Word32 pos)
 		cstr_SetPosition(pos);
 }
 //////////////////////////////////////////////
-void LDPUMA_DestroyRasterWnd()
+DPUMA_FUNC(void) LDPUMA_DestroyRasterWnd()
 {
 	if(DestroyRasterWnd)
 		DestroyRasterWnd();
@@ -697,7 +697,7 @@ void LDPUMA_CSTR_Update( )
 }
 
 //////////////////////////////////////////////
-Bool32 LDPUMA_SkipEx(Handle owner, Bool32 bIter, Bool32 bParent, Int32 nSign )
+DPUMA_FUNC(Bool32) LDPUMA_SkipEx(Handle owner, Bool32 bIter, Bool32 bParent, Int32 nSign )
 {
 	Bool32 rc = TRUE;
 	if(SkipEx)
@@ -713,7 +713,7 @@ Bool32 LDPUMA_OpenFile(Handle wnd, char * name )
 	return rc;
 }
 //////////////////////////////////////////////
-char * LDPUMA_GetFileName(Handle wnd)
+DPUMA_FUNC(char *) LDPUMA_GetFileName(Handle wnd)
 {
 	char * rc = "";
 	if(fGetFileName)
@@ -721,7 +721,7 @@ char * LDPUMA_GetFileName(Handle wnd)
 	return rc;
 }
 //////////////////////////////////////////////
-void *  LDPUMA_HandLayout( void * lpDIB, Word32 flags ,Point32 * p)
+DPUMA_FUNC(void *)  LDPUMA_HandLayout( void * lpDIB, Word32 flags ,Point32 * p)
 {
 	void * rc = NULL;
 	if(fHandLayout)
@@ -739,7 +739,7 @@ Bool32 LDPUMA_LockImage( Handle wnd, Bool32 bLock )
 return rc;
 }
 //////////////////////////////////////////////
-void LDPUMA_RegistryHelp( Handle owner, const char * lpstrHelp,Bool32 bAppend )
+DPUMA_FUNC(void) LDPUMA_RegistryHelp( Handle owner, const char * lpstrHelp,Bool32 bAppend )
 {
 	if(RegistryHelp)
 		RegistryHelp(owner,lpstrHelp,bAppend);
@@ -753,19 +753,19 @@ Bool32 LDPUMA_SaveFile( void * lpDIB,char * lpFileName,Word32 nFormat )
 	return rc;
 }
 //////////////////////////////////////////////
-void LDPUMA_ProgressStart( void )
+DPUMA_FUNC(void) LDPUMA_ProgressStart( void )
 {
 	if(fProgressStart)
 		fProgressStart();
 }
 //////////////////////////////////////////////
-void LDPUMA_ProgressFinish( void )
+DPUMA_FUNC(void) LDPUMA_ProgressFinish( void )
 {
 	if(fProgressFinish)
 		fProgressFinish();
 }
 //////////////////////////////////////////////
-Bool32 LDPUMA_ProgressStep(Word32 step, const char * name, Word32 percent)
+DPUMA_FUNC(Bool32) LDPUMA_ProgressStep(Word32 step, const char * name, Word32 percent)
 {
 	Bool32 rc = TRUE;
 	if(fProgressStep)
@@ -773,7 +773,7 @@ Bool32 LDPUMA_ProgressStep(Word32 step, const char * name, Word32 percent)
 	return rc;
 }
 //////////////////////////////////////////////
-Bool32 LDPUMA_SetConsoleProperty(Bool32 bold,Bool32 italic,
+DPUMA_FUNC(Bool32) LDPUMA_SetConsoleProperty(Bool32 bold,Bool32 italic,
 	Bool32 strikeout,Bool32 underline,Int32 height,	Int32 offset,	Word32 textcolor,
 	Int32 charset,	const char * name )
 {
@@ -836,7 +836,7 @@ Handle LDPUMA_TimeStamp(const char * name,Handle hTimer)
 return (Handle) clock1;
 }
 //////////////////////////////////////////////
-Handle LDPUMA_GetWindowHandle(const char * name)
+DPUMA_FUNC(Handle) LDPUMA_GetWindowHandle(const char * name)
 {
 	Handle rc = NULL;
 	if(fGetWindowHandle)
@@ -852,13 +852,13 @@ Handle LDPUMA_GetPrevSkipOwner()
 	return rc;
 }
 //////////////////////////////////////////////
-void LDPUMA_ConsoleClear(Int32 beforeline)
+DPUMA_FUNC(void) LDPUMA_ConsoleClear(Int32 beforeline)
 {
 	if(fConsoleClear)
 		fConsoleClear(beforeline);
 }
 //////////////////////////////////////////////
-Int32 LDPUMA_ConsoleGetCurLine()
+DPUMA_FUNC(Int32) LDPUMA_ConsoleGetCurLine()
 {
 	Int32 rc = -1;
 	if(fConsoleGetCurLine)
@@ -866,7 +866,7 @@ Int32 LDPUMA_ConsoleGetCurLine()
 return rc;
 }
 //////////////////////////////////////////////
-Bool32 LDPUMA_SetFileName(Handle wnd,char * lpName)
+DPUMA_FUNC(Bool32) LDPUMA_SetFileName(Handle wnd,char * lpName)
 {
 	Bool32 rc = FALSE;
 	if(fSetFileName)
diff --git a/cuneiform_src/Kern/hhh/rshelllines.h b/cuneiform_src/Kern/hhh/rshelllines.h
index 52e4b13..ae582c7 100644
--- a/cuneiform_src/Kern/hhh/rshelllines.h
+++ b/cuneiform_src/Kern/hhh/rshelllines.h
@@ -66,10 +66,16 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //
 //
 
-#ifdef RSHELLLINES_EXPORTS
-#define RSHELLLINES_API __declspec(dllexport)
+#ifndef __GLOBUS_H
+#include "globus.h"
+#endif
+
+#ifdef __RSHELLLINES_EXPORTS__
+  #define RSHELLLINES_FUNC  FUN_EXPO
+  #define RSHELLLINES_CLASS CLASS_EXPO
 #else
-#define RSHELLLINES_API __declspec(dllimport)
+  #define RSHELLLINES_FUNC  FUN_IMPO
+  #define RSHELLLINES_CLASS CLASS_IMPO
 #endif
 
 #define step             8
diff --git a/cuneiform_src/Kern/hhh/rshelllinescom.h b/cuneiform_src/Kern/hhh/rshelllinescom.h
index 586c089..c529432 100644
--- a/cuneiform_src/Kern/hhh/rshelllinescom.h
+++ b/cuneiform_src/Kern/hhh/rshelllinescom.h
@@ -65,10 +65,16 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //	Implemented: by B.M. Shahverdiev
 //
 //
-#ifdef RSHELLLINES_EXPORTS
-#define RSHELLLINES_API __declspec(dllexport)
+#ifndef __GLOBUS_H
+#include "globus.h"
+#endif
+
+#ifdef __RSHELLLINES_EXPORTS__
+  #define RSHELLLINES_FUNC  FUN_EXPO
+  #define RSHELLLINES_CLASS CLASS_EXPO
 #else
-#define RSHELLLINES_API __declspec(dllimport)
+  #define RSHELLLINES_FUNC  FUN_IMPO
+  #define RSHELLLINES_CLASS CLASS_IMPO
 #endif
 
 
@@ -149,7 +155,7 @@ typedef struct group_lines_mass
 
 /////////////////////////////////////////////////////////////////////////////
 // Interval's class
-/*class RSHELLLINES_API CInterval:public CObject
+/*class RSHELLLINES_CLASS CInterval:public CObject
 {
 public:
   CInterval(){}
@@ -161,7 +167,7 @@ public:
 */
 /////////////////////////////////////////////////////////////////////////////
 // Rect32's class
-/*class RSHELLLINES_API CPRect32:public CObject
+/*class RSHELLLINES_CLASS CPRect32:public CObject
 {
 public:
   CPRect32(){m_flag=TRUE;};
@@ -174,7 +180,7 @@ public:
 
 /////////////////////////////////////////////////////////////////////////////
 // Event's class
-/*class RSHELLLINES_API CEvent:public CObject
+/*class RSHELLLINES_CLASS CEvent:public CObject
 {
 public:
   CEvent();
@@ -197,7 +203,7 @@ public:
 */
 /////////////////////////////////////////////////////////////////////////////
 // Stripe's class
-/*class RSHELLLINES_API CStripe:public CObject
+/*class RSHELLLINES_CLASS CStripe:public CObject
 {
 public:
   CStripe();
@@ -219,7 +225,7 @@ public:
 //#define DirecteDownwards      1
 //#define DirectedForBothSides  2
 /*
-class RSHELLLINES_API CCutPoint:public CObject
+class RSHELLLINES_CLASS CCutPoint:public CObject
 {
 public:
   CCutPoint();
@@ -242,7 +248,7 @@ public:
 #define  MiddleHasCrossPoint        2
 #define  EndPointHasCrossPoint      4
 /*
-class RSHELLLINES_API CComponent:public CObject
+class RSHELLLINES_CLASS CComponent:public CObject
 {
 public:
   CComponent();
@@ -255,7 +261,7 @@ public:
 */
 /////////////////////////////////////////////////////////////////////////////
 // СDotLines's class
-/*class RSHELLLINES_API CDotLine:public CObject
+/*class RSHELLLINES_CLASS CDotLine:public CObject
 {
 public:
   CDotLine();
@@ -293,13 +299,13 @@ public:
 // Line's class
 //class CPageLines;
 /*
-class RSHELLLINES_API CLine:public CObject
+class RSHELLLINES_CLASS CLine:public CObject
 {
 public:
   CLine();
  ~CLine();
 */
-RSHELLLINES_API void            InitLine(DLine* linedata);  //
+RSHELLLINES_FUNC(void)            InitLine(DLine* linedata);  //
 /*  Word32          Flags;                   // common use info
   Word32          Tail;                    // остаток линии за таблицой
   Point32         BeginPoint;
@@ -368,7 +374,7 @@ RSHELLLINES_API void            InitLine(DLine* linedata);  //
   void            SetWidth(CLINE_handle line, DLine* pCLine); //
   void            SetLength(CLINE_handle line, DLine* pCLine); //
   void            SetComponentsCrossPointFlag(CLINE_handle line, DLine* pCLine); //
-RSHELLLINES_API Bool32          CheckSeparationPoints(CLINE_handle hLine, CLINE_handle hComp = NULL); //
+RSHELLLINES_FUNC(Bool32)          CheckSeparationPoints(CLINE_handle hLine, CLINE_handle hComp = NULL); //
 
   void            SetSeparationPoints(CLINE_handle hContainer, CLINE_handle hLine, DLine* pCLine, Bool32 PLines); //
   void            SetCutPoints(CLINE_handle line, DLine* pCLine); //
@@ -386,7 +392,7 @@ RSHELLLINES_API Bool32          CheckSeparationPoints(CLINE_handle hLine, CLINE_
 */
 /////////////////////////////////////////////////////////////////////////////
 // CGroupOfFriendLines class
-/*class RSHELLLINES_API CGroupOfFriendLines:public CObject
+/*class RSHELLLINES_CLASS CGroupOfFriendLines:public CObject
 {
 public:
   CGroupOfFriendLines();
@@ -408,7 +414,7 @@ public:
 */
 /////////////////////////////////////////////////////////////////////////////
 // CGroupOfExtensibleLines class
-/*class RSHELLLINES_API CGroupOfExtensibleLines:public CObject
+/*class RSHELLLINES_CLASS CGroupOfExtensibleLines:public CObject
 {
 public:
   CGroupOfExtensibleLines();
@@ -423,7 +429,7 @@ public:
 */
 /////////////////////////////////////////////////////////////////////////////
 // PageLine's class
-/*class RSHELLLINES_API CPageLines:public CObject
+/*class RSHELLLINES_CLASS CPageLines:public CObject
 {
 public:
   CPageLines();
@@ -471,42 +477,42 @@ public:
   BOOL                     CheckLinesDegreeForExtensible(CLINE_handle hLine, GLM* hGroupEx); //
 //  BOOL                     CheckGroupsForExtensible(void);
 
-RSHELLLINES_API void       FindFriendLines(CLINE_handle hContainer, GLM* friendlinesmass); //
-RSHELLLINES_API void       DrawFriendLines(CLINE_handle hContainer, GLM* friendlinesmass); //
+RSHELLLINES_FUNC(void)       FindFriendLines(CLINE_handle hContainer, GLM* friendlinesmass); //
+RSHELLLINES_FUNC(void)       DrawFriendLines(CLINE_handle hContainer, GLM* friendlinesmass); //
 
-RSHELLLINES_API void       FindGroupOfExtensibleLines(CLINE_handle hContainer, GLM* friendlinesmass, Bool32 IfDrawResult); //
+RSHELLLINES_FUNC(void)       FindGroupOfExtensibleLines(CLINE_handle hContainer, GLM* friendlinesmass, Bool32 IfDrawResult); //
   void                     DrawGroupOfExtensibleLines(CLINE_handle hContainer, GLM* hGroup, int CountLines); //
 
-RSHELLLINES_API void       FindLosedVerticalLines(CLINE_handle hContainer, GLM* friendlinesmass, CLINE_handle hFictContainer, int CountLines, Bool32 IfDrawResult); //
-RSHELLLINES_API void       DrawLosedVerticalLines(GLM* friendlinesmass, int CountLines); //
+RSHELLLINES_FUNC(void)       FindLosedVerticalLines(CLINE_handle hContainer, GLM* friendlinesmass, CLINE_handle hFictContainer, int CountLines, Bool32 IfDrawResult); //
+RSHELLLINES_FUNC(void)       DrawLosedVerticalLines(GLM* friendlinesmass, int CountLines); //
 
 //  void                     GetBigComps(Handle hCComp);
-RSHELLLINES_API void       DrawBigComps(CLINE_handle hContainer); //
+RSHELLLINES_FUNC(void)       DrawBigComps(CLINE_handle hContainer); //
 
-RSHELLLINES_API void       FindDotLines(Handle hCCOM,Handle hCPAGE, CLINE_handle hContainer); //
+RSHELLLINES_FUNC(void)       FindDotLines(Handle hCCOM,Handle hCPAGE, CLINE_handle hContainer); //
   BOOL                     GetDotComps(Handle hCCOM,Handle hCPAGE, HANDLE &hbuf,LPSTR &pDotBuffer, Int32 *CountDotComps); //
   void                     FindChains(Handle hCCOM, LPSTR &pDotBuffer, Int32 CountDotComps, CLINE_handle hContainer); //
   void                     CheckChains(Handle hCCOM, LPSTR &pDotBuffer, Int32 BegIndex,Int32 CountDots, CLINE_handle hContainer); //
   void                     AddNewDotLines(Handle hCCOM, LPSTR &pDotBuffer, Int32 CountDots, CLINE_handle hContainer); //
   void                     UnionBreakup(CLINE_handle hContainer); //
 
-RSHELLLINES_API void       SetLinesAndCompsRelationship(CLINE_handle hContainer, CLINE_handle hFictContainer); //
+RSHELLLINES_FUNC(void)       SetLinesAndCompsRelationship(CLINE_handle hContainer, CLINE_handle hFictContainer); //
 
-RSHELLLINES_API void       DrowAllLines(CLINE_handle hContainer, Handle hDrowAllLines); //
+RSHELLLINES_FUNC(void)       DrowAllLines(CLINE_handle hContainer, Handle hDrowAllLines); //
 /*  int                      SetImage(Handle hCPage);
 };
 */
-RSHELLLINES_API Bool32 FindLineFrag(CLINE_handle processedline, Bool32 OnlyPosyAndStat, Bool32 Is2ndPath, CLINE_handle hContainer, Bool32 IfNeedFragment = TRUE, Bool32 IfStraightFrag = FALSE); //
-RSHELLLINES_API Bool32 FindLineAttr(CLINE_handle line, DLine* pCLine, Bool32 AbleSeeOldAttr); //
-RSHELLLINES_API Bool32 GetNextPartOfLine(CLINE_handle hContainer, CLINE_handle hLine/*, CPDLine oldlinedata*/); //
+RSHELLLINES_FUNC(Bool32) FindLineFrag(CLINE_handle processedline, Bool32 OnlyPosyAndStat, Bool32 Is2ndPath, CLINE_handle hContainer, Bool32 IfNeedFragment = TRUE, Bool32 IfStraightFrag = FALSE); //
+RSHELLLINES_FUNC(Bool32) FindLineAttr(CLINE_handle line, DLine* pCLine, Bool32 AbleSeeOldAttr); //
+RSHELLLINES_FUNC(Bool32) GetNextPartOfLine(CLINE_handle hContainer, CLINE_handle hLine/*, CPDLine oldlinedata*/); //
 
 void                       AddLine2Group(GLM* hGroup, CLINE_handle hLine);
-RSHELLLINES_API void       DeleteGroup(GLM* hGroup, int CountLines);
-RSHELLLINES_API void       DrawFragsForAllLines(CLINE_handle hContainer, Handle hDebugDrawAllLineFragments);
-RSHELLLINES_API Int32	   CountBlackRaster(CPDLine pLine, CPDLine pLineExt, Bool32 IsHor, Int32 delta, Handle hDrawRaster);
-RSHELLLINES_API Bool32     SetExtLines(CLINE_handle hExtContainer, CLINE_handle hContainer, CLINE_handle* hLinesMass, Int32 CountLines);
-RSHELLLINES_API Bool32     RSL_CorrectDoubleLines(CLINE_handle hLine1, CLINE_handle hLine2);
-RSHELLLINES_API Bool32     RSL_SplitLine(CLINE_handle hLine, CLINE_handle hContainer);
+RSHELLLINES_FUNC(void)       DeleteGroup(GLM* hGroup, int CountLines);
+RSHELLLINES_FUNC(void)       DrawFragsForAllLines(CLINE_handle hContainer, Handle hDebugDrawAllLineFragments);
+RSHELLLINES_FUNC(Int32)	   CountBlackRaster(CPDLine pLine, CPDLine pLineExt, Bool32 IsHor, Int32 delta, Handle hDrawRaster);
+RSHELLLINES_FUNC(Bool32)     SetExtLines(CLINE_handle hExtContainer, CLINE_handle hContainer, CLINE_handle* hLinesMass, Int32 CountLines);
+RSHELLLINES_FUNC(Bool32)     RSL_CorrectDoubleLines(CLINE_handle hLine1, CLINE_handle hLine2);
+RSHELLLINES_FUNC(Bool32)     RSL_SplitLine(CLINE_handle hLine, CLINE_handle hContainer);
 
 //possible return values for RSL_VerifyShortLine
 #define RSL_IS_LINE 0
@@ -517,8 +523,8 @@ RSHELLLINES_API Bool32     RSL_SplitLine(CLINE_handle hLine, CLINE_handle hConta
 
 #define MAX_CROSS_POINTS 4
 
-RSHELLLINES_API Int32     RSL_VerifyShortLine(CPDLine pLine, Handle hCCOM, PAGEINFO* page_info, Word8 lang, Word8 debug_flags, Int32 *cross_point = NULL);
-RSHELLLINES_API Bool   SL_GetRaster(Rect32* rect, Word8** ppData, PAGEINFO* page_info/*, CIMAGEBITMAPINFOHEADER* image_info*/); //
+RSHELLLINES_FUNC(Int32)     RSL_VerifyShortLine(CPDLine pLine, Handle hCCOM, PAGEINFO* page_info, Word8 lang, Word8 debug_flags, Int32 *cross_point = NULL);
+RSHELLLINES_FUNC(Bool)   SL_GetRaster(Rect32* rect, Word8** ppData, PAGEINFO* page_info/*, CIMAGEBITMAPINFOHEADER* image_info*/); //
 
 
 //}
diff --git a/cuneiform_src/Kern/hrkint/cpu.h b/cuneiform_src/Kern/hrkint/cpu.h
index babad3f..5cdbbcb 100644
--- a/cuneiform_src/Kern/hrkint/cpu.h
+++ b/cuneiform_src/Kern/hrkint/cpu.h
@@ -59,13 +59,23 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #include "decl.h"
 
+#ifndef __GLOBUS_H
+#include "globus.h"
+#endif
+
+#ifdef __CPU__
+  #define CPU_FUNC  FUN_EXPO
+#else
+  #define CPU_FUNC  FUN_IMPO
+#endif
+
 #ifdef __cplusplus
     extern "C" {
 #endif
 
-__declspec( dllimport ) int GetCPUName(void);
-__declspec( dllimport ) int Get_MMX(void);
-__declspec( dllimport ) int Get_XMM(void);
+CPU_FUNC(int) GetCPUName(void);
+CPU_FUNC(int) Get_MMX(void);
+CPU_FUNC(int) Get_XMM(void);
 #ifdef __cplusplus
     }
 #endif
diff --git a/cuneiform_src/Kern/include/winfuncs.h b/cuneiform_src/Kern/include/winfuncs.h
index 313b9db..f005c92 100644
--- a/cuneiform_src/Kern/include/winfuncs.h
+++ b/cuneiform_src/Kern/include/winfuncs.h
@@ -34,68 +34,78 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #else
 
+#ifndef __GLOBUS_H
+#include "globus.h"
+#endif
+
+#ifdef __WINDUMMY__
+  #define WINDUMMY_FUNC  FUN_EXPO
+#else
+  #define WINDUMMY_FUNC  FUN_IMPO
+#endif
+
 #include<wchar.h>
 
-int LoadString(HINSTANCE hInstance, UINT uID, LPTSTR lpBuffer, int nBufferMax);
+WINDUMMY_FUNC(int) LoadString(HINSTANCE hInstance, UINT uID, LPTSTR lpBuffer, int nBufferMax);
 
-int CreateDirectory(const char *dir, void *dummy);
-DWORD GetTempPath(DWORD nBufferLength, LPTSTR lpBuffer);
-int RemoveDirectory(const char *d);
-void* GlobalAlloc(UINT uFlags, int dwBytes);
-HGLOBAL GlobalFree(void *f);
-void* GlobalReAlloc(void* hMem, int dwBytes, UINT uFlags);
-int GetTempFileName(LPCTSTR lpPathName, LPCTSTR lpPrefixString,
+WINDUMMY_FUNC(int) CreateDirectory(const char *dir, void *dummy);
+WINDUMMY_FUNC(DWORD) GetTempPath(DWORD nBufferLength, LPTSTR lpBuffer);
+WINDUMMY_FUNC(int) RemoveDirectory(const char *d);
+WINDUMMY_FUNC(void*) GlobalAlloc(UINT uFlags, int dwBytes);
+WINDUMMY_FUNC(HGLOBAL) GlobalFree(void *f);
+WINDUMMY_FUNC(void*) GlobalReAlloc(void* hMem, int dwBytes, UINT uFlags);
+WINDUMMY_FUNC(int) GetTempFileName(LPCTSTR lpPathName, LPCTSTR lpPrefixString,
                     UINT uUnique, LPTSTR lpTempFileName);
-int GetLastError();
+WINDUMMY_FUNC(int) GetLastError();
 
-DWORD GetModuleFileName(HMODULE hModule, LPTSTR lpFilename, DWORD nSize);
-BOOL CloseHandle(HANDLE hObject);
-HANDLE CreateFile(LPCTSTR lpFileName, DWORD dwDesiredAccess,
+WINDUMMY_FUNC(DWORD) GetModuleFileName(HMODULE hModule, LPTSTR lpFilename, DWORD nSize);
+WINDUMMY_FUNC(BOOL) CloseHandle(HANDLE hObject);
+WINDUMMY_FUNC(HANDLE) CreateFile(LPCTSTR lpFileName, DWORD dwDesiredAccess,
 DWORD dwShareMode, void* lpSecurityAttributes,
 DWORD dwCreationDisposition, DWORD dwFlagsAndAttributes, HANDLE hTemplateFile);
 
-HWND FindWindow(LPCTSTR lpClassName, LPCTSTR lpWindowName);
-UINT RegisterWindowMessage(LPCTSTR lpString);
+WINDUMMY_FUNC(HWND) FindWindow(LPCTSTR lpClassName, LPCTSTR lpWindowName);
+WINDUMMY_FUNC(UINT) RegisterWindowMessage(LPCTSTR lpString);
 
-int _findclose(long handle);
-long _findfirst(const char *filespec, struct _finddata_t *fileinfo);
-int _findnext(long handle, struct _finddata_t *fileinfo);
-long _tell(int handle);
+WINDUMMY_FUNC(int) _findclose(long handle);
+WINDUMMY_FUNC(long) _findfirst(const char *filespec, struct _finddata_t *fileinfo);
+WINDUMMY_FUNC(int) _findnext(long handle, struct _finddata_t *fileinfo);
+WINDUMMY_FUNC(long) _tell(int handle);
 
-BOOL GetComputerName(LPTSTR lpBuffer, long unsigned int *lpnSize);
+WINDUMMY_FUNC(BOOL) GetComputerName(LPTSTR lpBuffer, long unsigned int *lpnSize);
 
-LONG RegOpenKeyEx(HKEY hKey, LPCTSTR lpSubKey, DWORD ulOptions,
+WINDUMMY_FUNC(LONG) RegOpenKeyEx(HKEY hKey, LPCTSTR lpSubKey, DWORD ulOptions,
   REGSAM samDesired, PHKEY phkResult);
-LONG RegQueryValueEx(HKEY hKey, LPCTSTR lpValueName,
+WINDUMMY_FUNC(LONG) RegQueryValueEx(HKEY hKey, LPCTSTR lpValueName,
 LPDWORD lpReserved, LPDWORD lpType, LPBYTE lpData,
 LPDWORD lpcbData);
 
-BOOL GetClientRect(HWND hWnd, LPRECT lpRect);
-BOOL WritePrivateProfileString(LPCTSTR lpAppName,
+WINDUMMY_FUNC(BOOL) GetClientRect(HWND hWnd, LPRECT lpRect);
+WINDUMMY_FUNC(BOOL) WritePrivateProfileString(LPCTSTR lpAppName,
         LPCTSTR lpKeyName, LPCTSTR lpString, LPCTSTR lpFileName);
-DWORD GetPrivateProfileString(LPCTSTR lpAppName, LPCTSTR lpKeyName,
+WINDUMMY_FUNC(DWORD) GetPrivateProfileString(LPCTSTR lpAppName, LPCTSTR lpKeyName,
 LPCTSTR lpDefault, LPTSTR lpReturnedString, DWORD nSize, LPCTSTR lpFileName);
-UINT GetPrivateProfileInt(LPCTSTR lpAppName,
+WINDUMMY_FUNC(UINT) GetPrivateProfileInt(LPCTSTR lpAppName,
   LPCTSTR lpKeyName, INT nDefault, LPCTSTR lpFileName);
 
-int WideCharToMultiByte(UINT CodePage, DWORD dwFlags, const wchar_t *lpWideCharStr,
+WINDUMMY_FUNC(int) WideCharToMultiByte(UINT CodePage, DWORD dwFlags, const wchar_t *lpWideCharStr,
   int cchWideChar, LPSTR lpMultiByteStr, int cbMultiByte,
   LPCSTR lpDefaultChar, LPBOOL lpUsedDefaultChar);
 
-BOOL ShowWindow(HWND hWnd, int nCmdShow);
-
-long _filelength(int fd);
-long _msize(void *memblock);
-int _access(const char *filename, int mode);
-BOOL SetWindowText(HWND hWnd,LPCTSTR lpString);
-int ReleaseDC(HWND hWnd, HDC hDC);
-BOOL IsIconic(HWND hWnd);
-HDC GetDC(HWND hWnd);
-BOOL EndPaint(HWND hWnd, ...);
-HDC BeginPaint(HWND hwnd,...);
-LRESULT SendMessage(HWND hWnd, UINT Msg, WPARAM wParam, LPARAM lParam);
-void strlwr(char *foo);
-HWND CreateWindow(
+WINDUMMY_FUNC(BOOL) ShowWindow(HWND hWnd, int nCmdShow);
+
+WINDUMMY_FUNC(long) _filelength(int fd);
+WINDUMMY_FUNC(long) _msize(void *memblock);
+WINDUMMY_FUNC(int) _access(const char *filename, int mode);
+WINDUMMY_FUNC(BOOL) SetWindowText(HWND hWnd,LPCTSTR lpString);
+WINDUMMY_FUNC(int) ReleaseDC(HWND hWnd, HDC hDC);
+WINDUMMY_FUNC(BOOL) IsIconic(HWND hWnd);
+WINDUMMY_FUNC(HDC) GetDC(HWND hWnd);
+WINDUMMY_FUNC(BOOL) EndPaint(HWND hWnd, ...);
+WINDUMMY_FUNC(HDC) BeginPaint(HWND hwnd,...);
+WINDUMMY_FUNC(LRESULT) SendMessage(HWND hWnd, UINT Msg, WPARAM wParam, LPARAM lParam);
+WINDUMMY_FUNC(void) strlwr(char *foo);
+WINDUMMY_FUNC(HWND) CreateWindow(
     LPCTSTR lpClassName,
     LPCTSTR lpWindowName,
     DWORD dwStyle,
@@ -109,62 +119,62 @@ HWND CreateWindow(
     LPVOID lpParam
 );
 
-HGDIOBJ SelectObject(HDC hdc, HGDIOBJ hgdiobj);
-LPTSTR lstrcat(LPTSTR lpString1, LPTSTR lpString2);
-int lstrlen(LPCTSTR lpString);
-int lstrcmp(LPCTSTR lpString1, LPCTSTR lpString2);
-LPTSTR lstrcpy(LPTSTR lpString1, LPCTSTR lpString2);
-int wsprintf(LPTSTR lpOut, LPCTSTR lpFmt, ...);
-int lstrcmpi(LPCTSTR lpString1, LPCTSTR lpString2);
+WINDUMMY_FUNC(HGDIOBJ) SelectObject(HDC hdc, HGDIOBJ hgdiobj);
+WINDUMMY_FUNC(LPTSTR) lstrcat(LPTSTR lpString1, LPTSTR lpString2);
+WINDUMMY_FUNC(int) lstrlen(LPCTSTR lpString);
+WINDUMMY_FUNC(int) lstrcmp(LPCTSTR lpString1, LPCTSTR lpString2);
+WINDUMMY_FUNC(LPTSTR) lstrcpy(LPTSTR lpString1, LPCTSTR lpString2);
+WINDUMMY_FUNC(int) wsprintf(LPTSTR lpOut, LPCTSTR lpFmt, ...);
+WINDUMMY_FUNC(int) lstrcmpi(LPCTSTR lpString1, LPCTSTR lpString2);
 
-BOOL DeleteObject(HGDIOBJ hObject);
+WINDUMMY_FUNC(BOOL) DeleteObject(HGDIOBJ hObject);
 
 #define GetGValue(rgb) ((BYTE) (((WORD) (rgb)) >> 8))
 #define GetBValue(rgb) ((BYTE) ((rgb) >> 16))
 #define GetRValue(rgb) ((BYTE) (rgb))
 
-HWND GetFocus();
-int MessageBox(HWND hWnd, LPCTSTR lpText, LPCTSTR lpCaption, UINT uType);
+WINDUMMY_FUNC(HWND) GetFocus();
+WINDUMMY_FUNC(int) MessageBox(HWND hWnd, LPCTSTR lpText, LPCTSTR lpCaption, UINT uType);
 
-int WINAPI GlobalSize(HGLOBAL hMem);
-LPVOID GlobalLock(HGLOBAL hMem);
-BOOL GlobalUnlock(HGLOBAL hMem);
-BOOL IsBadWritePtr(LPVOID lp, int ucb);
-void OutputDebugString(LPCTSTR lpOutputString);
-BOOL SetRect(LPRECT lprc, int xLeft, int yTop,
+WINDUMMY_FUNC(int) WINAPI GlobalSize(HGLOBAL hMem);
+WINDUMMY_FUNC(LPVOID) GlobalLock(HGLOBAL hMem);
+WINDUMMY_FUNC(BOOL) GlobalUnlock(HGLOBAL hMem);
+WINDUMMY_FUNC(BOOL) IsBadWritePtr(LPVOID lp, int ucb);
+WINDUMMY_FUNC(void) OutputDebugString(LPCTSTR lpOutputString);
+WINDUMMY_FUNC(BOOL) SetRect(LPRECT lprc, int xLeft, int yTop,
 int xRight, int yBottom);
-BOOL PtInRect(const RECT *lprc, POINT pt);
-BOOL IntersectRect(LPRECT lprcDst, const RECT *lprcSrc1, const RECT *lprcSrc2);
-BOOL UnionRect(LPRECT lprcDst, const RECT *lprcSrc1,const RECT *lprcSrc2);
+WINDUMMY_FUNC(BOOL) PtInRect(const RECT *lprc, POINT pt);
+WINDUMMY_FUNC(BOOL) IntersectRect(LPRECT lprcDst, const RECT *lprcSrc1, const RECT *lprcSrc2);
+WINDUMMY_FUNC(BOOL) UnionRect(LPRECT lprcDst, const RECT *lprcSrc1,const RECT *lprcSrc2);
 
-HWND GetActiveWindow();
-HFONT CreateFont(int nHeight, int nWidth, int nEscapement,
+WINDUMMY_FUNC(HWND) GetActiveWindow();
+WINDUMMY_FUNC(HFONT) CreateFont(int nHeight, int nWidth, int nEscapement,
 int nOrientation, int fnWeight, DWORD fdwItalic, DWORD fdwUnderline,
 DWORD fdwStrikeOut, DWORD fdwCharSet, DWORD fdwOutputPrecision,
 DWORD fdwClipPrecision, DWORD fdwQuality, DWORD fdwPitchAndFamily,
 LPCTSTR lpszFace);
-BOOL GetTextExtentPoint32(HDC hdc, LPCTSTR lpString, int c, LPSIZE lpSize);
-BOOL EnumWindows(WNDENUMPROC lpEnumFunc, LPARAM lParam);
-int GetWindowText(HWND hWnd, LPTSTR lpString, int nMaxCount);
+WINDUMMY_FUNC(BOOL) GetTextExtentPoint32(HDC hdc, LPCTSTR lpString, int c, LPSIZE lpSize);
+WINDUMMY_FUNC(BOOL) EnumWindows(WNDENUMPROC lpEnumFunc, LPARAM lParam);
+WINDUMMY_FUNC(int) GetWindowText(HWND hWnd, LPTSTR lpString, int nMaxCount);
 
-HMODULE LoadLibrary(LPCTSTR lpFileName);
-BOOL FreeLibrary(HMODULE hModule);
-void* GetProcAddress(HMODULE hModule, LPCSTR lpProcName);
+WINDUMMY_FUNC(HMODULE) LoadLibrary(LPCTSTR lpFileName);
+WINDUMMY_FUNC(BOOL) FreeLibrary(HMODULE hModule);
+WINDUMMY_FUNC(void*) GetProcAddress(HMODULE hModule, LPCSTR lpProcName);
 
-HGDIOBJ GetStockObject(int fnObject);
-BOOL IsWindowVisible(HWND hWnd);
-LRESULT DefWindowProc(HWND hWnd, UINT Msg, WPARAM wParam,
+WINDUMMY_FUNC(HGDIOBJ) GetStockObject(int fnObject);
+WINDUMMY_FUNC(BOOL) IsWindowVisible(HWND hWnd);
+WINDUMMY_FUNC(LRESULT) DefWindowProc(HWND hWnd, UINT Msg, WPARAM wParam,
     LPARAM lParam);
-LONG GetWindowLong(HWND hWnd, int nIndex);
-BOOL RegisterClass(const WNDCLASS *lpWndClass);
-HMODULE GetModuleHandle(LPCTSTR lpModuleName);
-HICON LoadIcon(HINSTANCE hInstance, LPCTSTR lpIconName);
-
-long filelength(int fd);
-int LoadCursor(HINSTANCE hInstance, LPCTSTR lpCursorName);
-BOOL Rectangle(HDC hdc,
+WINDUMMY_FUNC(LONG) GetWindowLong(HWND hWnd, int nIndex);
+WINDUMMY_FUNC(BOOL) RegisterClass(const WNDCLASS *lpWndClass);
+WINDUMMY_FUNC(HMODULE) GetModuleHandle(LPCTSTR lpModuleName);
+WINDUMMY_FUNC(HICON) LoadIcon(HINSTANCE hInstance, LPCTSTR lpIconName);
+
+WINDUMMY_FUNC(long) filelength(int fd);
+WINDUMMY_FUNC(int) LoadCursor(HINSTANCE hInstance, LPCTSTR lpCursorName);
+WINDUMMY_FUNC(BOOL) Rectangle(HDC hdc,
   int nLeftRect, int nTopRect, int nRightRect, int nBottomRect);
-char* _strupr(char*s);
+WINDUMMY_FUNC(char*) _strupr(char*s);
 
 #endif /* not WIN32 */
 
@@ -176,22 +186,22 @@ char* _strupr(char*s);
 extern "C" {
 #endif
 
-int open_data_file(const char *basename, int mode);
-int data_file_exists(const char *basename);
+WINDUMMY_FUNC(int) open_data_file(const char *basename, int mode);
+WINDUMMY_FUNC(int) data_file_exists(const char *basename);
 
-void split_path(const char *fname,
+WINDUMMY_FUNC(void) split_path(const char *fname,
         char *file_path,
         char *basename,
         char *ext);
 
-void make_path(char *opath,
+WINDUMMY_FUNC(void) make_path(char *opath,
         const char *dir,
         const char *basename,
         const char *ext);
 
-void winpath_to_internal(char *p);
+WINDUMMY_FUNC(void) winpath_to_internal(char *p);
 
-unsigned int curr_dir(unsigned int bsize, char* buf);
+WINDUMMY_FUNC(unsigned int) curr_dir(unsigned int bsize, char* buf);
 
 #ifdef __cplusplus
 }
diff --git a/cuneiform_src/Kern/itigerole/sources/tiger/h/globus.h b/cuneiform_src/Kern/itigerole/sources/tiger/h/globus.h
index 273f9da..f1d7eb4 100644
--- a/cuneiform_src/Kern/itigerole/sources/tiger/h/globus.h
+++ b/cuneiform_src/Kern/itigerole/sources/tiger/h/globus.h
@@ -138,6 +138,19 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
             #define CDECL      __cdecl
          #endif
       #endif
+   #elif    defined( __GNUC_MINOR__ ) && defined(HAVE_GCCVISIBILITY) /* GNU Compiler Suite ******/
+         #define CLA_IMPO
+         #define CLA_EXPO __attribute__ ((visibility("default")))
+         #define FUN_IMPO__
+         #define __FUN_IMPO
+         #define FUN_EXPO__ __attribute__ ((visibility("default")))
+         #define __FUN_EXPO
+         #ifndef PASCAL
+            #define PASCAL __attribute__ ((stdcall))
+         #endif
+         #ifndef CDECL
+            #define CDECL __attribute__ ((cdecl))
+         #endif
    #else /* unknown compiler ******************************************/
          #define CLA_IMPO
          #define CLA_EXPO
diff --git a/cuneiform_src/Kern/leo/CMakeLists.txt b/cuneiform_src/Kern/leo/CMakeLists.txt
index 8855eb8..c8623f7 100644
--- a/cuneiform_src/Kern/leo/CMakeLists.txt
+++ b/cuneiform_src/Kern/leo/CMakeLists.txt
@@ -12,6 +12,7 @@ src/leo_prn.c
 src/leo_size.c
 src/leo_stic.c
 )
+set_property(TARGET leo PROPERTY COMPILE_DEFINITIONS __LEO__)
 
 target_link_libraries(leo
 cpu
diff --git a/cuneiform_src/Kern/lns32/CMakeLists.txt b/cuneiform_src/Kern/lns32/CMakeLists.txt
index 96c92e1..205cfd0 100644
--- a/cuneiform_src/Kern/lns32/CMakeLists.txt
+++ b/cuneiform_src/Kern/lns32/CMakeLists.txt
@@ -24,6 +24,7 @@ src/smooth.cpp
 src/sweeper.cpp
 src/tgreader.cpp
 )
+set_property(TARGET lns32 PROPERTY COMPILE_DEFINITIONS __LNS__)
 
 target_link_libraries(lns32 cline std32 windummy)
 
diff --git a/cuneiform_src/Kern/loc/CMakeLists.txt b/cuneiform_src/Kern/loc/CMakeLists.txt
index c74ff67..8136dd1 100644
--- a/cuneiform_src/Kern/loc/CMakeLists.txt
+++ b/cuneiform_src/Kern/loc/CMakeLists.txt
@@ -7,6 +7,7 @@ src/loc.c
 src/locompmn.c
 src/v0compgl.c
 )
+set_property(TARGET loc PROPERTY COMPILE_DEFINITIONS __LOC__)
 
 target_link_libraries(loc windummy)
 
diff --git a/cuneiform_src/Kern/loc/src/locompmn.c b/cuneiform_src/Kern/loc/src/locompmn.c
index 938ca49..ca9a640 100644
--- a/cuneiform_src/Kern/loc/src/locompmn.c
+++ b/cuneiform_src/Kern/loc/src/locompmn.c
@@ -59,6 +59,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <stdlib.h>
 #include "struct.h"
 #include "v1comp.h"
+#include "loc.h"
 
 BWSS *locomp_seglist(Word8* raster, BWSS *bwsp, BWSS *bwe, Int32 height, Int32 width);
 MN * c_locomp (Word8* raster, Int32 bw, Int32 h, Int16 upper, Int16 left);
@@ -101,7 +102,7 @@ static void new_line_cont();
 static void merge_line();
 static void dead_line();
 
-MN * c_locomp (Word8* raster, Int32 bw, Int32 h, Int16 upper, Int16 left)
+LOC_FUNC(MN *) c_locomp (Word8* raster, Int32 bw, Int32 h, Int16 upper, Int16 left)
 {
  lineno = upper-1; rast_lc = left;
  if (setjmp(locomp_err)) return NULL;
diff --git a/cuneiform_src/Kern/mmx/CMakeLists.txt b/cuneiform_src/Kern/mmx/CMakeLists.txt
index 2eb0cda..d3e5b71 100644
--- a/cuneiform_src/Kern/mmx/CMakeLists.txt
+++ b/cuneiform_src/Kern/mmx/CMakeLists.txt
@@ -7,6 +7,7 @@ src/mmx_net.c
 src/mmx_r35.c
 src/mmxmain.c
 )
+set_property(TARGET mmx PROPERTY COMPILE_DEFINITIONS __MMX__)
 
 if(WIN32)
   install(TARGETS mmx ARCHIVE DESTINATION ${LIBDIR})
diff --git a/cuneiform_src/Kern/msk/CMakeLists.txt b/cuneiform_src/Kern/msk/CMakeLists.txt
index 3e1441f..0f9ebba 100644
--- a/cuneiform_src/Kern/msk/CMakeLists.txt
+++ b/cuneiform_src/Kern/msk/CMakeLists.txt
@@ -1,6 +1,7 @@
 include_directories(src)
 
 add_library(msk ${LIBTYPE} src/msk.c)
+set_property(TARGET msk PROPERTY COMPILE_DEFINITIONS __MSK__)
 
 target_link_libraries(msk mmx windummy)
 
diff --git a/cuneiform_src/Kern/pass2/CMakeLists.txt b/cuneiform_src/Kern/pass2/CMakeLists.txt
index 528ca02..5dcdcca 100644
--- a/cuneiform_src/Kern/pass2/CMakeLists.txt
+++ b/cuneiform_src/Kern/pass2/CMakeLists.txt
@@ -7,6 +7,7 @@ src/p2_rstr.c
 src/p2_test.c
 src/p2_tools.c
 )
+set_property(TARGET pass2 PROPERTY COMPILE_DEFINITIONS __P2__)
 
 target_link_libraries(pass2 cstr fon std32)
 
diff --git a/cuneiform_src/Kern/puma/CMakeLists.txt b/cuneiform_src/Kern/puma/CMakeLists.txt
index d2b44a5..87b6c45 100644
--- a/cuneiform_src/Kern/puma/CMakeLists.txt
+++ b/cuneiform_src/Kern/puma/CMakeLists.txt
@@ -13,6 +13,7 @@ main/mymem.cpp
 main/puma.cpp 
 main/dll.cpp 
 )
+set_property(TARGET puma PROPERTY COMPILE_DEFINITIONS __PUMA__)
 
 target_link_libraries(puma
 ced
diff --git a/cuneiform_src/Kern/puma/c/partlayout.cpp b/cuneiform_src/Kern/puma/c/partlayout.cpp
index 09848cb..a508225 100644
--- a/cuneiform_src/Kern/puma/c/partlayout.cpp
+++ b/cuneiform_src/Kern/puma/c/partlayout.cpp
@@ -58,8 +58,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 	#include <windows.h>
 #endif
 
-#define __PUMA__
-
 #include <stdio.h>
 #include <string.h>
 
diff --git a/cuneiform_src/Kern/puma/main/dll.cpp b/cuneiform_src/Kern/puma/main/dll.cpp
index 6f47bc4..9af736b 100644
--- a/cuneiform_src/Kern/puma/main/dll.cpp
+++ b/cuneiform_src/Kern/puma/main/dll.cpp
@@ -61,8 +61,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //
 // dll_cpage.cpp :
 // ============================================================================
-#define __PUMA__
-
 /*#include <windows.h>*/
 #include <string.h>
 #include <stdio.h>
diff --git a/cuneiform_src/Kern/puma/main/puma.cpp b/cuneiform_src/Kern/puma/main/puma.cpp
index 21ed85f..82d6b69 100644
--- a/cuneiform_src/Kern/puma/main/puma.cpp
+++ b/cuneiform_src/Kern/puma/main/puma.cpp
@@ -55,7 +55,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
 #define __PUMA_CPP__
-#define __PUMA__
 //#include <windows.h>
 //#include <stdio.h>
 #include <string>
diff --git a/cuneiform_src/Kern/r35/CMakeLists.txt b/cuneiform_src/Kern/r35/CMakeLists.txt
index 4b8cf0a..f2a59cb 100644
--- a/cuneiform_src/Kern/r35/CMakeLists.txt
+++ b/cuneiform_src/Kern/r35/CMakeLists.txt
@@ -1,5 +1,6 @@
 add_library(r35 ${LIBTYPE}
 src/r35.c)
+set_property(TARGET r35 PROPERTY COMPILE_DEFINITIONS __R35__)
 
 target_link_libraries(r35 mmx std32 windummy)
 
diff --git a/cuneiform_src/Kern/rbal/CMakeLists.txt b/cuneiform_src/Kern/rbal/CMakeLists.txt
index 1009fa6..b84fe86 100644
--- a/cuneiform_src/Kern/rbal/CMakeLists.txt
+++ b/cuneiform_src/Kern/rbal/CMakeLists.txt
@@ -10,6 +10,7 @@ src/linpon.c
 src/linutil.c
 src/statsearchbl.cpp
 )
+set_property(TARGET rbal PROPERTY COMPILE_DEFINITIONS __RBAL__)
 
 target_link_libraries(rbal cstr evn32)
 
diff --git a/cuneiform_src/Kern/rblock/CMakeLists.txt b/cuneiform_src/Kern/rblock/CMakeLists.txt
index a97464a..926ba48 100644
--- a/cuneiform_src/Kern/rblock/CMakeLists.txt
+++ b/cuneiform_src/Kern/rblock/CMakeLists.txt
@@ -49,6 +49,7 @@ sources/c/semain.c
 sources/c/sestring.c
 sources/c/sevbreak.c
 )
+set_property(TARGET rblock PROPERTY COMPILE_DEFINITIONS __RBLOCK__)
 
 target_link_libraries(rblock cpage loc exc cstr cline rstr)
 
diff --git a/cuneiform_src/Kern/rblock/sources/main/_dll.cpp b/cuneiform_src/Kern/rblock/sources/main/_dll.cpp
index 7dcd26f..93dea6c 100644
--- a/cuneiform_src/Kern/rblock/sources/main/_dll.cpp
+++ b/cuneiform_src/Kern/rblock/sources/main/_dll.cpp
@@ -60,8 +60,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //
 // dll_cpage.cpp :
 // ============================================================================
-#define __RBLOCK__
-
 /*#include <windows.h>*/
 #include <setjmp.h>
 
diff --git a/cuneiform_src/Kern/rblock/sources/main/_rblock.cpp b/cuneiform_src/Kern/rblock/sources/main/_rblock.cpp
index 2143841..3deed74 100644
--- a/cuneiform_src/Kern/rblock/sources/main/_rblock.cpp
+++ b/cuneiform_src/Kern/rblock/sources/main/_rblock.cpp
@@ -54,8 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __RBLOCK__
-
 /*#include <windows.h>*/
 #include <setjmp.h>
 #include <assert.h>
diff --git a/cuneiform_src/Kern/rcorrkegl/CMakeLists.txt b/cuneiform_src/Kern/rcorrkegl/CMakeLists.txt
index b99be5a..a7b6958 100644
--- a/cuneiform_src/Kern/rcorrkegl/CMakeLists.txt
+++ b/cuneiform_src/Kern/rcorrkegl/CMakeLists.txt
@@ -6,6 +6,7 @@ src/cpp/cor_kegl.cpp
 src/cpp/rcorrkegl.cpp
 src/cpp/stdafx.cpp
 )
+set_property(TARGET rcorrkegl PROPERTY COMPILE_DEFINITIONS __RCORRKEGL__)
 
 target_link_libraries(rcorrkegl cstr cpage)
 
diff --git a/cuneiform_src/Kern/rcutp/CMakeLists.txt b/cuneiform_src/Kern/rcutp/CMakeLists.txt
index fa824b6..9df9a05 100644
--- a/cuneiform_src/Kern/rcutp/CMakeLists.txt
+++ b/cuneiform_src/Kern/rcutp/CMakeLists.txt
@@ -9,6 +9,7 @@ sources/main/rcutp_func.cpp
 sources/main/rcutp_main.cpp
 sources/main/rcutp_mainfunc.cpp
 )
+set_property(TARGET rcutp PROPERTY COMPILE_DEFINITIONS __RCUTP__)
 
 target_link_libraries(rcutp ccom windummy)
 
diff --git a/cuneiform_src/Kern/rdib/CMakeLists.txt b/cuneiform_src/Kern/rdib/CMakeLists.txt
index 0e14d9a..6ad4267 100644
--- a/cuneiform_src/Kern/rdib/CMakeLists.txt
+++ b/cuneiform_src/Kern/rdib/CMakeLists.txt
@@ -1,6 +1,7 @@
 include_directories(BEFORE sources)
 
 add_library(rdib ${LIBTYPE} sources/cpp/CTDIB.cpp)
+set_property(TARGET rdib PROPERTY COMPILE_DEFINITIONS __RDIB__)
 
 if(WIN32)
   install(TARGETS rdib ARCHIVE DESTINATION ${LIBDIR})
diff --git a/cuneiform_src/Kern/rdib/sources/cpp/CTDIB.cpp b/cuneiform_src/Kern/rdib/sources/cpp/CTDIB.cpp
index e75e70c..e3f175e 100644
--- a/cuneiform_src/Kern/rdib/sources/cpp/CTDIB.cpp
+++ b/cuneiform_src/Kern/rdib/sources/cpp/CTDIB.cpp
@@ -57,9 +57,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 // CTDIB.cpp: implementation of the CTDIB class.
 //
 //////////////////////////////////////////////////////////////////////
-#define __RDIB__
-//////////////////////////////////////////////////////////////////////////////////////////////////////////////
-
 #include <memory.h>
 #include <stdlib.h>
 #include "ctdib.h"
diff --git a/cuneiform_src/Kern/rfrmt/CMakeLists.txt b/cuneiform_src/Kern/rfrmt/CMakeLists.txt
index 1a13cb5..df12c17 100644
--- a/cuneiform_src/Kern/rfrmt/CMakeLists.txt
+++ b/cuneiform_src/Kern/rfrmt/CMakeLists.txt
@@ -19,6 +19,7 @@ sources/main/util.cpp
 sources/main/util_lst.cpp
 sources/main/util_spl.cpp
 )
+set_property(TARGET rfrmt PROPERTY COMPILE_DEFINITIONS __RFRMT__)
 
 target_link_libraries(rfrmt
 ced
diff --git a/cuneiform_src/Kern/rfrmt/sources/main/frmt.cpp b/cuneiform_src/Kern/rfrmt/sources/main/frmt.cpp
index b84a467..86f6424 100644
--- a/cuneiform_src/Kern/rfrmt/sources/main/frmt.cpp
+++ b/cuneiform_src/Kern/rfrmt/sources/main/frmt.cpp
@@ -69,7 +69,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <unistd.h>
 #include "stdafx.h"
 #include "resource.h"
-#define __RFRMT__
 #include "rfrmt.h"
 #include "mymem.h"
 #include "rtffragrect.h"
diff --git a/cuneiform_src/Kern/rfrmt/sources/main/frmtdll.cpp b/cuneiform_src/Kern/rfrmt/sources/main/frmtdll.cpp
index 6c2de96..1bb7998 100644
--- a/cuneiform_src/Kern/rfrmt/sources/main/frmtdll.cpp
+++ b/cuneiform_src/Kern/rfrmt/sources/main/frmtdll.cpp
@@ -63,7 +63,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 /*#include <windows.h>*/
 #include "resource.h"
-#define __RFRMT__
 #define __RFRMT_EXTERN__
 #include "rfrmt.h"
 #include "dpuma.h"
diff --git a/cuneiform_src/Kern/rimage/CMakeLists.txt b/cuneiform_src/Kern/rimage/CMakeLists.txt
index db55982..0edab79 100644
--- a/cuneiform_src/Kern/rimage/CMakeLists.txt
+++ b/cuneiform_src/Kern/rimage/CMakeLists.txt
@@ -16,6 +16,7 @@ sources/main/crturner.cpp
 sources/main/dll.cpp
 sources/main/rprogressor.cpp
 )
+set_property(TARGET rimage PROPERTY COMPILE_DEFINITIONS __RIMAGE__)
 
 target_link_libraries(rimage
 cfio
diff --git a/cuneiform_src/Kern/rimage/sources/main/criimage.cpp b/cuneiform_src/Kern/rimage/sources/main/criimage.cpp
index bac8ed0..a5fa5ec 100644
--- a/cuneiform_src/Kern/rimage/sources/main/criimage.cpp
+++ b/cuneiform_src/Kern/rimage/sources/main/criimage.cpp
@@ -54,8 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __RIMAGE__
-
 #include "resource.h"
 #include "criimage.h"
 #include "crimemory.h"
diff --git a/cuneiform_src/Kern/rimage/sources/main/dll.cpp b/cuneiform_src/Kern/rimage/sources/main/dll.cpp
index 5d32295..1b7fed1 100644
--- a/cuneiform_src/Kern/rimage/sources/main/dll.cpp
+++ b/cuneiform_src/Kern/rimage/sources/main/dll.cpp
@@ -60,8 +60,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //
 // dll_cpage.cpp :
 // ============================================================================
-#define __RIMAGE__
-
 /*#include <windows.h>*/
 
 #include "resource.h"
diff --git a/cuneiform_src/Kern/rline/CMakeLists.txt b/cuneiform_src/Kern/rline/CMakeLists.txt
index 496e68f..7613275 100644
--- a/cuneiform_src/Kern/rline/CMakeLists.txt
+++ b/cuneiform_src/Kern/rline/CMakeLists.txt
@@ -6,6 +6,7 @@ sources/findlostlines.cpp
 sources/newline.cpp
 sources/rline.cpp
 )
+set_property(TARGET rline PROPERTY COMPILE_DEFINITIONS __RLINE__)
 
 target_link_libraries(rline
 cimage
diff --git a/cuneiform_src/Kern/rline/sources/dll.cpp b/cuneiform_src/Kern/rline/sources/dll.cpp
index 946145d..92e27a1 100644
--- a/cuneiform_src/Kern/rline/sources/dll.cpp
+++ b/cuneiform_src/Kern/rline/sources/dll.cpp
@@ -58,8 +58,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //
 // ============================================================================
 
-#define __RLINE__
-
 #include <stdlib.h>
 #include "stdafx.h"
 #include "resource.h"
diff --git a/cuneiform_src/Kern/rline/sources/findlostlines.cpp b/cuneiform_src/Kern/rline/sources/findlostlines.cpp
index 5a1c3e9..a7c0e39 100644
--- a/cuneiform_src/Kern/rline/sources/findlostlines.cpp
+++ b/cuneiform_src/Kern/rline/sources/findlostlines.cpp
@@ -54,8 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __RLINE__
-
 #ifdef _DEBUG
 #define SOURCE_FILE_NAME "FindLostLines.cpp"
 #endif
diff --git a/cuneiform_src/Kern/rline/sources/newline.cpp b/cuneiform_src/Kern/rline/sources/newline.cpp
index 123f46a..d8950a4 100644
--- a/cuneiform_src/Kern/rline/sources/newline.cpp
+++ b/cuneiform_src/Kern/rline/sources/newline.cpp
@@ -54,8 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __RLINE__
-
 #include <stdlib.h>
 #include <stdio.h>
 #include <math.h>
diff --git a/cuneiform_src/Kern/rline/sources/rline.cpp b/cuneiform_src/Kern/rline/sources/rline.cpp
index b027705..323cd62 100644
--- a/cuneiform_src/Kern/rline/sources/rline.cpp
+++ b/cuneiform_src/Kern/rline/sources/rline.cpp
@@ -58,8 +58,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //
 // ============================================================================
 
-#define __RLINE__
-
 #include <stdlib.h>
 /*#include <windows.h>*/
 #include <memory.h>
diff --git a/cuneiform_src/Kern/rling/CMakeLists.txt b/cuneiform_src/Kern/rling/CMakeLists.txt
index e21d2f3..03b9b2b 100644
--- a/cuneiform_src/Kern/rling/CMakeLists.txt
+++ b/cuneiform_src/Kern/rling/CMakeLists.txt
@@ -37,6 +37,7 @@ sources/c/spelspec.c
 sources/c/udictini.c
 sources/c/udictuti.c
 )
+set_property(TARGET rling PROPERTY COMPILE_DEFINITIONS __RLING__)
 
 target_link_libraries(rling cstr rlings)
 
diff --git a/cuneiform_src/Kern/rling/sources/CMakeLists.txt b/cuneiform_src/Kern/rling/sources/CMakeLists.txt
index e262e74..ad8986e 100644
--- a/cuneiform_src/Kern/rling/sources/CMakeLists.txt
+++ b/cuneiform_src/Kern/rling/sources/CMakeLists.txt
@@ -31,6 +31,7 @@ c/spelspec.c
 c/udictini.c
 c/udictuti.c
 )
+set_property(TARGET rlings PROPERTY COMPILE_DEFINITIONS __RLINGS__)
 
 target_link_libraries(rlings cstr windummy)
 
diff --git a/cuneiform_src/Kern/rling/sources/cpp/crling.cpp b/cuneiform_src/Kern/rling/sources/cpp/crling.cpp
index edc2d10..b83de71 100644
--- a/cuneiform_src/Kern/rling/sources/cpp/crling.cpp
+++ b/cuneiform_src/Kern/rling/sources/cpp/crling.cpp
@@ -54,12 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#ifndef RLING_SECONDARY
-  #define __RLING__
-#else
-  #define __RLINGS__
-#endif
-
 #include "resource.h"
 #include "rlcontrol.h"
 #include "crling.h"
diff --git a/cuneiform_src/Kern/rling/sources/cpp/dll.cpp b/cuneiform_src/Kern/rling/sources/cpp/dll.cpp
index d10c751..601a9f3 100644
--- a/cuneiform_src/Kern/rling/sources/cpp/dll.cpp
+++ b/cuneiform_src/Kern/rling/sources/cpp/dll.cpp
@@ -61,12 +61,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //
 // dll_cpage.cpp :
 // ============================================================================
-#ifndef RLING_SECONDARY
-  #define __RLING__
-#else
-  #define __RLINGS__
-#endif
-
 /*#include <windows.h>*/
 
 #include "resource.h"
diff --git a/cuneiform_src/Kern/rmarker/CMakeLists.txt b/cuneiform_src/Kern/rmarker/CMakeLists.txt
index 3bccf03..7ae1761 100644
--- a/cuneiform_src/Kern/rmarker/CMakeLists.txt
+++ b/cuneiform_src/Kern/rmarker/CMakeLists.txt
@@ -13,6 +13,7 @@ shortverticallinesfilter.cpp
 ../usage/puma_err.cpp
 ../usage/un_buff.cpp
 )
+set_property(TARGET rmarker PROPERTY COMPILE_DEFINITIONS __RMARKER__)
 
 target_link_libraries(rmarker
 cfio
diff --git a/cuneiform_src/Kern/rneg/CMakeLists.txt b/cuneiform_src/Kern/rneg/CMakeLists.txt
index f92f270..e4fbac0 100644
--- a/cuneiform_src/Kern/rneg/CMakeLists.txt
+++ b/cuneiform_src/Kern/rneg/CMakeLists.txt
@@ -9,6 +9,7 @@ sources/src/negmain/recnegmain.cpp
 sources/src/negmain/rnegbase.cpp
 sources/src/searchneg/searchneg.cpp
 )
+set_property(TARGET rneg PROPERTY COMPILE_DEFINITIONS __RNEG__)
 
 target_link_libraries(rneg
 cimage
diff --git a/cuneiform_src/Kern/rneg/sources/src/negatest/negatestcell.cpp b/cuneiform_src/Kern/rneg/sources/src/negatest/negatestcell.cpp
index 8482e41..fbac69f 100644
--- a/cuneiform_src/Kern/rneg/sources/src/negatest/negatestcell.cpp
+++ b/cuneiform_src/Kern/rneg/sources/src/negatest/negatestcell.cpp
@@ -54,8 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __RNEG__
-
 #include "rneg.h"
 #include "recnegmain.h"
 #include "compat_defs.h"
diff --git a/cuneiform_src/Kern/rneg/sources/src/negmain/recnegmain.cpp b/cuneiform_src/Kern/rneg/sources/src/negmain/recnegmain.cpp
index 083e866..9d26670 100644
--- a/cuneiform_src/Kern/rneg/sources/src/negmain/recnegmain.cpp
+++ b/cuneiform_src/Kern/rneg/sources/src/negmain/recnegmain.cpp
@@ -54,8 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __RNEG__
-
 #include <string.h>
 #include "rneg.h"
 #include "recnegmain.h"
diff --git a/cuneiform_src/Kern/rneg/sources/src/negmain/rnegbase.cpp b/cuneiform_src/Kern/rneg/sources/src/negmain/rnegbase.cpp
index 541b547..e796c1f 100644
--- a/cuneiform_src/Kern/rneg/sources/src/negmain/rnegbase.cpp
+++ b/cuneiform_src/Kern/rneg/sources/src/negmain/rnegbase.cpp
@@ -54,8 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __RNEG__
-
 #include "rneg.h"
 #include "rnegbase.h"
 #include "compat_defs.h"
diff --git a/cuneiform_src/Kern/rout/CMakeLists.txt b/cuneiform_src/Kern/rout/CMakeLists.txt
index 162002f..f0f1bfa 100644
--- a/cuneiform_src/Kern/rout/CMakeLists.txt
+++ b/cuneiform_src/Kern/rout/CMakeLists.txt
@@ -13,6 +13,7 @@ src/tabletext.cpp
 src/text.cpp
 src/words.cpp
 )
+set_property(TARGET rout PROPERTY COMPILE_DEFINITIONS __ROUT__)
 
 target_link_libraries(rout ced cfio)
 
diff --git a/cuneiform_src/Kern/rout/src/rout_own.h b/cuneiform_src/Kern/rout/src/rout_own.h
index d626f1d..0006ee3 100644
--- a/cuneiform_src/Kern/rout/src/rout_own.h
+++ b/cuneiform_src/Kern/rout/src/rout_own.h
@@ -59,8 +59,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #pragma once
 
-#define __ROUT__
-
 #include "stdafx.h"
 #include "rout.h"
 #include "cfio.h"
diff --git a/cuneiform_src/Kern/rpic/CMakeLists.txt b/cuneiform_src/Kern/rpic/CMakeLists.txt
index cdeb49f..ff25871 100644
--- a/cuneiform_src/Kern/rpic/CMakeLists.txt
+++ b/cuneiform_src/Kern/rpic/CMakeLists.txt
@@ -12,6 +12,7 @@ sources/polypic.cpp
 sources/rectpic.cpp
 sources/rpic.cpp
 )
+set_property(TARGET rpic PROPERTY COMPILE_DEFINITIONS __RPIC__)
 
 target_link_libraries(rpic cpage ccom rstr)
 
diff --git a/cuneiform_src/Kern/rpic/sources/dll.cpp b/cuneiform_src/Kern/rpic/sources/dll.cpp
index 0a66be9..27337e5 100644
--- a/cuneiform_src/Kern/rpic/sources/dll.cpp
+++ b/cuneiform_src/Kern/rpic/sources/dll.cpp
@@ -58,8 +58,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //
 // ============================================================================
 
-#define __RPIC__
-
 /*# include <windows.h>*/
 #include "compat_defs.h"
 
diff --git a/cuneiform_src/Kern/rpic/sources/rpic.cpp b/cuneiform_src/Kern/rpic/sources/rpic.cpp
index 1cd635a..69a09f9 100644
--- a/cuneiform_src/Kern/rpic/sources/rpic.cpp
+++ b/cuneiform_src/Kern/rpic/sources/rpic.cpp
@@ -58,8 +58,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //
 // ============================================================================
 
-#define __RPIC__
-
 #define __RPIC_CPP__
 
 /*#include <windows.h>*/
diff --git a/cuneiform_src/Kern/rpstr/CMakeLists.txt b/cuneiform_src/Kern/rpstr/CMakeLists.txt
index fcf26b7..6ce6693 100644
--- a/cuneiform_src/Kern/rpstr/CMakeLists.txt
+++ b/cuneiform_src/Kern/rpstr/CMakeLists.txt
@@ -8,6 +8,7 @@ src/incl/cor_incl.c
 src/cor_spel.c 
 src/rpstr.c 
 src/vertstr.cpp)
+set_property(TARGET rpstr PROPERTY COMPILE_DEFINITIONS __RPSTR__)
 
 target_link_libraries(rpstr
 ccom
diff --git a/cuneiform_src/Kern/rreccom/CMakeLists.txt b/cuneiform_src/Kern/rreccom/CMakeLists.txt
index bdcb1b4..b4dbf95 100644
--- a/cuneiform_src/Kern/rreccom/CMakeLists.txt
+++ b/cuneiform_src/Kern/rreccom/CMakeLists.txt
@@ -5,6 +5,7 @@ src/cpp/alphabet.cpp
 src/cpp/recog.cpp
 src/cpp/rreccom.cpp
 )
+set_property(TARGET rreccom PROPERTY COMPILE_DEFINITIONS __RRECCOM__)
 
 target_link_libraries(rreccom ccom evn32 windummy)
 
diff --git a/cuneiform_src/Kern/rreccom/src/cpp/recog.cpp b/cuneiform_src/Kern/rreccom/src/cpp/recog.cpp
index 1ddd8ce..4090fc1 100644
--- a/cuneiform_src/Kern/rreccom/src/cpp/recog.cpp
+++ b/cuneiform_src/Kern/rreccom/src/cpp/recog.cpp
@@ -55,8 +55,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
 /*********************************************************************************************/
-#define __RRECCOM__
-
 //#define _USE_GRA_ 1
 /*********************************************************************************************/
 /*#include <windows.h>
diff --git a/cuneiform_src/Kern/rreccom/src/cpp/rreccom.cpp b/cuneiform_src/Kern/rreccom/src/cpp/rreccom.cpp
index 3de977f..75b8dad 100644
--- a/cuneiform_src/Kern/rreccom/src/cpp/rreccom.cpp
+++ b/cuneiform_src/Kern/rreccom/src/cpp/rreccom.cpp
@@ -55,8 +55,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
 /*********************************************************************************************/
-#define __RRECCOM__
-/*********************************************************************************************/
 #include <string.h>
 #include "stdafx.h"
 #include "rreccom.h"
diff --git a/cuneiform_src/Kern/rsadd/CMakeLists.txt b/cuneiform_src/Kern/rsadd/CMakeLists.txt
index fbafdbb..70d41da 100644
--- a/cuneiform_src/Kern/rsadd/CMakeLists.txt
+++ b/cuneiform_src/Kern/rsadd/CMakeLists.txt
@@ -1,6 +1,7 @@
 include_directories(BEFORE src)
 
 add_library(rsadd ${LIBTYPE} src/rsadd.c)
+set_property(TARGET rsadd PROPERTY COMPILE_DEFINITIONS __RSADD__)
 
 target_link_libraries(rsadd cstr)
 
diff --git a/cuneiform_src/Kern/rselstr/CMakeLists.txt b/cuneiform_src/Kern/rselstr/CMakeLists.txt
index 8c58c7c..48a74f8 100644
--- a/cuneiform_src/Kern/rselstr/CMakeLists.txt
+++ b/cuneiform_src/Kern/rselstr/CMakeLists.txt
@@ -40,6 +40,7 @@ sources/src/cpp/sestring.cpp
 sources/src/cpp/sevbreak.cpp
 sources/src/chstr/vertical/testforvertical.cpp
 )
+set_property(TARGET rselstr PROPERTY COMPILE_DEFINITIONS __RSELSTR__)
 
 target_link_libraries(rselstr
 cimage
diff --git a/cuneiform_src/Kern/rselstr/sources/src/chstr/addtocstr/puttocstr.cpp b/cuneiform_src/Kern/rselstr/sources/src/chstr/addtocstr/puttocstr.cpp
index 8e5cc6a..c93b336 100644
--- a/cuneiform_src/Kern/rselstr/sources/src/chstr/addtocstr/puttocstr.cpp
+++ b/cuneiform_src/Kern/rselstr/sources/src/chstr/addtocstr/puttocstr.cpp
@@ -54,8 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __RSELSTR__
-
 #include <string.h>
 #include <stdlib.h>
 #include "puttocstr.h"
diff --git a/cuneiform_src/Kern/rselstr/sources/src/chstr/cont/puttocont.cpp b/cuneiform_src/Kern/rselstr/sources/src/chstr/cont/puttocont.cpp
index 6c5913e..ce5fa0e 100644
--- a/cuneiform_src/Kern/rselstr/sources/src/chstr/cont/puttocont.cpp
+++ b/cuneiform_src/Kern/rselstr/sources/src/chstr/cont/puttocont.cpp
@@ -54,9 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __RSELSTR__
-
-
 #include "rselstr.h"
 #include "puttocont.h"
 
diff --git a/cuneiform_src/Kern/rselstr/sources/src/chstr/cutstr/chstr_cutstr.cpp b/cuneiform_src/Kern/rselstr/sources/src/chstr/cutstr/chstr_cutstr.cpp
index 1bc1a55..024548b 100644
--- a/cuneiform_src/Kern/rselstr/sources/src/chstr/cutstr/chstr_cutstr.cpp
+++ b/cuneiform_src/Kern/rselstr/sources/src/chstr/cutstr/chstr_cutstr.cpp
@@ -54,8 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __RSELSTR__
-
 #include "cutstr.h"
 #include "rselstr.h"
 #include <assert.h>
diff --git a/cuneiform_src/Kern/rselstr/sources/src/chstr/rotate/rotate.cpp b/cuneiform_src/Kern/rselstr/sources/src/chstr/rotate/rotate.cpp
index de1288e..563dd2d 100644
--- a/cuneiform_src/Kern/rselstr/sources/src/chstr/rotate/rotate.cpp
+++ b/cuneiform_src/Kern/rselstr/sources/src/chstr/rotate/rotate.cpp
@@ -54,7 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __RSELSTR__
 #include <string.h>
 #include "rotate.h"
 #include "rselstr.h"
diff --git a/cuneiform_src/Kern/rselstr/sources/src/chstr/vertical/testforvertical.cpp b/cuneiform_src/Kern/rselstr/sources/src/chstr/vertical/testforvertical.cpp
index 18ac177..aa323bc 100644
--- a/cuneiform_src/Kern/rselstr/sources/src/chstr/vertical/testforvertical.cpp
+++ b/cuneiform_src/Kern/rselstr/sources/src/chstr/vertical/testforvertical.cpp
@@ -54,8 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __RSELSTR__
-
 #include <stdlib.h>
 #include <string.h>
 #include "puttocstr.h"
diff --git a/cuneiform_src/Kern/rselstr/sources/src/cpp/cutstr.cpp b/cuneiform_src/Kern/rselstr/sources/src/cpp/cutstr.cpp
index d50126b..05bcda1 100644
--- a/cuneiform_src/Kern/rselstr/sources/src/cpp/cutstr.cpp
+++ b/cuneiform_src/Kern/rselstr/sources/src/cpp/cutstr.cpp
@@ -54,7 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __RSELSTR__
 #include<math.h>
 #include<time.h>
 #include "rselstr.h"
diff --git a/cuneiform_src/Kern/rselstr/sources/src/rselstrmain/rselstrbase.cpp b/cuneiform_src/Kern/rselstr/sources/src/rselstrmain/rselstrbase.cpp
index b179e60..e18cea7 100644
--- a/cuneiform_src/Kern/rselstr/sources/src/rselstrmain/rselstrbase.cpp
+++ b/cuneiform_src/Kern/rselstr/sources/src/rselstrmain/rselstrbase.cpp
@@ -54,7 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __RSELSTR__
 /*#include <windows.h>*/
 #include <setjmp.h>
 
diff --git a/cuneiform_src/Kern/rselstr/sources/src/rselstrmain/rselstrmain.cpp b/cuneiform_src/Kern/rselstr/sources/src/rselstrmain/rselstrmain.cpp
index 2196efc..80e66aa 100644
--- a/cuneiform_src/Kern/rselstr/sources/src/rselstrmain/rselstrmain.cpp
+++ b/cuneiform_src/Kern/rselstr/sources/src/rselstrmain/rselstrmain.cpp
@@ -54,9 +54,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define __RSELSTR__
-
-
 # include <setjmp.h>
 # include <string.h>
 #include<math.h>
diff --git a/cuneiform_src/Kern/rshelllines/CMakeLists.txt b/cuneiform_src/Kern/rshelllines/CMakeLists.txt
index b2fde95..d761dac 100644
--- a/cuneiform_src/Kern/rshelllines/CMakeLists.txt
+++ b/cuneiform_src/Kern/rshelllines/CMakeLists.txt
@@ -9,6 +9,7 @@ src/rshelllines.cpp
 #src/RShellLines.rc
 src/stdafx.cpp
 )
+set_property(TARGET rshelllines PROPERTY COMPILE_DEFINITIONS __RSHELLLINES_EXPORTS__)
 
 target_link_libraries(rshelllines
 cfio
diff --git a/cuneiform_src/Kern/rshelllines/src/rshelllines.cpp b/cuneiform_src/Kern/rshelllines/src/rshelllines.cpp
index 63aa2c5..16f0178 100644
--- a/cuneiform_src/Kern/rshelllines/src/rshelllines.cpp
+++ b/cuneiform_src/Kern/rshelllines/src/rshelllines.cpp
@@ -176,7 +176,7 @@ Bool32 AboutLines (PRSPreProcessImage Image, Bool32 *BadScan, Int32 *ScanQual);
 
 /////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 //                     FindLineFrag
-RSHELLLINES_API Bool32 FindLineFrag(CLINE_handle processedline, Bool32 OnlyPosyAndStat,
+RSHELLLINES_FUNC(Bool32) FindLineFrag(CLINE_handle processedline, Bool32 OnlyPosyAndStat,
 									Bool32 Is2ndPath, CLINE_handle hContainer,
 									Bool32 IfNeedFragment, Bool32 IfStraightFrag)
 {
@@ -185,7 +185,7 @@ RSHELLLINES_API Bool32 FindLineFrag(CLINE_handle processedline, Bool32 OnlyPosyA
 
 /////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 //                     FindLineAttr
-RSHELLLINES_API Bool32 FindLineAttr(CLINE_handle line, DLine* pCLine, Bool32 AbleSeeOldAttr)
+RSHELLLINES_FUNC(Bool32) FindLineAttr(CLINE_handle line, DLine* pCLine, Bool32 AbleSeeOldAttr)
 {
     //SetWidth(line, pCLine);
     return TRUE;
@@ -575,12 +575,12 @@ Bool32 CompareRasterParts(CPDLine pLine, LPSTR pSourceRaster, Bool32 CheckSerif)
     return TRUE;
 }
 
-RSHELLLINES_API Bool32 RSL_CorrectDoubleLines(CLINE_handle hLine1, CLINE_handle hLine2)
+RSHELLLINES_FUNC(Bool32) RSL_CorrectDoubleLines(CLINE_handle hLine1, CLINE_handle hLine2)
 {
 	return TRUE;
 }
 
-RSHELLLINES_API Bool32     RSL_SplitLine(CLINE_handle hLine, CLINE_handle hContainer)
+RSHELLLINES_FUNC(Bool32)     RSL_SplitLine(CLINE_handle hLine, CLINE_handle hContainer)
 {
 	return TRUE;
 }
@@ -662,7 +662,7 @@ Handle DrawLineFragments(CLINE_handle line, CPDLine pCLine, Handle HndMyWindow,
     return NULL;
 }
 
-RSHELLLINES_API void DrawFriendLines(CLINE_handle hContainer, GLM* friendlinesmass)
+RSHELLLINES_FUNC(void) DrawFriendLines(CLINE_handle hContainer, GLM* friendlinesmass)
 {
 }
 
@@ -670,58 +670,58 @@ void DrawGroupOfExtensibleLines(CLINE_handle hContainer, GLM* hGroup, int CountL
 {
 }
 
-RSHELLLINES_API void DrawLosedVerticalLines(GLM* friendlinesmass, int CountLines)
+RSHELLLINES_FUNC(void) DrawLosedVerticalLines(GLM* friendlinesmass, int CountLines)
 {
 }
 
-RSHELLLINES_API void DrowAllLines(CLINE_handle hContainer, Handle hDrowAllLines)
+RSHELLLINES_FUNC(void) DrowAllLines(CLINE_handle hContainer, Handle hDrowAllLines)
 {
 }
 
-RSHELLLINES_API void DrawBigComps(CLINE_handle hContainer)
+RSHELLLINES_FUNC(void) DrawBigComps(CLINE_handle hContainer)
 {
 }
 
-RSHELLLINES_API void DrawFragsForAllLines(CLINE_handle hContainer, Handle hDebugDrawAllLineFragments)
+RSHELLLINES_FUNC(void) DrawFragsForAllLines(CLINE_handle hContainer, Handle hDebugDrawAllLineFragments)
 {
 }
 
-RSHELLLINES_API void InitLine(DLine* linedata)
+RSHELLLINES_FUNC(void) InitLine(DLine* linedata)
 {
 }
 
-RSHELLLINES_API void FindDotLines(Handle hCCOM,Handle hCPAGE, CLINE_handle hContainer)
+RSHELLLINES_FUNC(void) FindDotLines(Handle hCCOM,Handle hCPAGE, CLINE_handle hContainer)
 {
 }
 
-RSHELLLINES_API Bool32 CheckSeparationPoints(CLINE_handle hLine, CLINE_handle hComp)
+RSHELLLINES_FUNC(Bool32) CheckSeparationPoints(CLINE_handle hLine, CLINE_handle hComp)
 {
     return FALSE;
 }
 
-RSHELLLINES_API Bool SL_GetRaster(Rect32* rect, Word8** ppData, PAGEINFO* page_info)
+RSHELLLINES_FUNC(Bool) SL_GetRaster(Rect32* rect, Word8** ppData, PAGEINFO* page_info)
 {
     return TRUE;
 }
 
-RSHELLLINES_API void SetLinesAndCompsRelationship(CLINE_handle hContainer, CLINE_handle hFictContainer)
+RSHELLLINES_FUNC(void) SetLinesAndCompsRelationship(CLINE_handle hContainer, CLINE_handle hFictContainer)
 {
 }
 
-RSHELLLINES_API void FindFriendLines(CLINE_handle hContainer, GLM* friendlinesmass)
+RSHELLLINES_FUNC(void) FindFriendLines(CLINE_handle hContainer, GLM* friendlinesmass)
 {
 }
 
-RSHELLLINES_API void FindGroupOfExtensibleLines(CLINE_handle hContainer, GLM* friendlinesmass, Bool32 IfDrawResult)
+RSHELLLINES_FUNC(void) FindGroupOfExtensibleLines(CLINE_handle hContainer, GLM* friendlinesmass, Bool32 IfDrawResult)
 {
 }
 
-RSHELLLINES_API Int32 RSL_VerifyShortLine(CPDLine pLine, Handle hCCOM, PAGEINFO* page_info, Word8 lang, Word8 debug_flags, Int32 *cross_point)
+RSHELLLINES_FUNC(Int32) RSL_VerifyShortLine(CPDLine pLine, Handle hCCOM, PAGEINFO* page_info, Word8 lang, Word8 debug_flags, Int32 *cross_point)
 {
     return 0;
 }
 
-RSHELLLINES_API Bool32 GetNextPartOfLine(CLINE_handle hContainer, CLINE_handle hLine)
+RSHELLLINES_FUNC(Bool32) GetNextPartOfLine(CLINE_handle hContainer, CLINE_handle hLine)
 {
     return FALSE;
 }
diff --git a/cuneiform_src/Kern/rstr/CMakeLists.txt b/cuneiform_src/Kern/rstr/CMakeLists.txt
index 702cafe..15a0dc9 100644
--- a/cuneiform_src/Kern/rstr/CMakeLists.txt
+++ b/cuneiform_src/Kern/rstr/CMakeLists.txt
@@ -88,6 +88,7 @@ src/sweeper.c
 src/tm.c
 src/tools.c
 src/ukr.c)
+set_property(TARGET rstr PROPERTY COMPILE_DEFINITIONS __RSTR__)
 
 target_link_libraries(rstr
 ccom
diff --git a/cuneiform_src/Kern/rstuff/CMakeLists.txt b/cuneiform_src/Kern/rstuff/CMakeLists.txt
index 9e2d151..7103dd3 100644
--- a/cuneiform_src/Kern/rstuff/CMakeLists.txt
+++ b/cuneiform_src/Kern/rstuff/CMakeLists.txt
@@ -25,6 +25,7 @@ sources/main/rstuff.cpp
 ../usage/puma_err.cpp
 ../usage/un_buff.cpp
 )
+set_property(TARGET rstuff PROPERTY COMPILE_DEFINITIONS __RSTUFF__)
 
 target_link_libraries(rstuff
 cfio
diff --git a/cuneiform_src/Kern/rstuff/sources/main/dll.cpp b/cuneiform_src/Kern/rstuff/sources/main/dll.cpp
index 31f2e0a..a70586f 100644
--- a/cuneiform_src/Kern/rstuff/sources/main/dll.cpp
+++ b/cuneiform_src/Kern/rstuff/sources/main/dll.cpp
@@ -64,8 +64,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 /*  Содержание :  Интерфейс библиотеки                                        */
 /*  Назначение :                                                              */
 /*----------------------------------------------------------------------------*/
-#define __RSTUFF__
-
 /*#include <windows.h>*/
 #include "resource.h"
 #include "rstuff.h"
diff --git a/cuneiform_src/Kern/rverline/CMakeLists.txt b/cuneiform_src/Kern/rverline/CMakeLists.txt
index 4424520..d4ce834 100644
--- a/cuneiform_src/Kern/rverline/CMakeLists.txt
+++ b/cuneiform_src/Kern/rverline/CMakeLists.txt
@@ -12,6 +12,7 @@ src/service/vl_data.cpp
 src/service/vl_rule.cpp
 src/service/vl_snap.cpp
 )
+set_property(TARGET rverline PROPERTY COMPILE_DEFINITIONS __RVERLINE__)
 
 target_link_libraries(rverline
 cimage
diff --git a/cuneiform_src/Kern/rverline/src/root/vl_kern.cpp b/cuneiform_src/Kern/rverline/src/root/vl_kern.cpp
index 3829fa5..a29d397 100644
--- a/cuneiform_src/Kern/rverline/src/root/vl_kern.cpp
+++ b/cuneiform_src/Kern/rverline/src/root/vl_kern.cpp
@@ -61,8 +61,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 /*  Содержание :  Содержательные функции библиотеки "RVERLINE"                */
 /*  Назначение :  Стандартный диалог с другими библиотеками                   */
 /*----------------------------------------------------------------------------*/
-#define __RVERLINE__
-
 #include <stdio.h>
 /*#include <windows.h>*/
 #include <string.h>
diff --git a/cuneiform_src/Kern/rverline/src/root/vl_resid.cpp b/cuneiform_src/Kern/rverline/src/root/vl_resid.cpp
index 759a4e1..1256509 100644
--- a/cuneiform_src/Kern/rverline/src/root/vl_resid.cpp
+++ b/cuneiform_src/Kern/rverline/src/root/vl_resid.cpp
@@ -61,8 +61,6 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 /*  Содержание :  Общие функции библиотеки "RVERLINE"                         */
 /*  Назначение :  Стандартный диалог с другими библиотеками                   */
 /*----------------------------------------------------------------------------*/
-#define __RVERLINE__
-
 #include <stdio.h>
 #include <string.h>
 /*#include <windows.h>*/
diff --git a/cuneiform_src/Kern/smetric/CMakeLists.txt b/cuneiform_src/Kern/smetric/CMakeLists.txt
index 10742b6..29b25fd 100644
--- a/cuneiform_src/Kern/smetric/CMakeLists.txt
+++ b/cuneiform_src/Kern/smetric/CMakeLists.txt
@@ -10,6 +10,7 @@ src/ugol.cpp
 ../usage/puma_comm.cpp
 ../usage/un_buff.cpp
 )
+set_property(TARGET smetric PROPERTY COMPILE_DEFINITIONS __SMETRIC__)
 
 target_link_libraries(smetric windummy)
 
diff --git a/cuneiform_src/Kern/std/CMakeLists.txt b/cuneiform_src/Kern/std/CMakeLists.txt
index fa6e8b1..c441773 100644
--- a/cuneiform_src/Kern/std/CMakeLists.txt
+++ b/cuneiform_src/Kern/std/CMakeLists.txt
@@ -17,6 +17,7 @@ src/stdprt.cpp
 src/stdsort.cpp
 src/stdwapi.cpp
 )
+set_property(TARGET std32 PROPERTY COMPILE_DEFINITIONS __STD__)
 
 target_link_libraries(std32 windummy)
 
diff --git a/cuneiform_src/Kern/windummy.c b/cuneiform_src/Kern/windummy.c
index 1ae8672..d38a4a1 100644
--- a/cuneiform_src/Kern/windummy.c
+++ b/cuneiform_src/Kern/windummy.c
@@ -60,7 +60,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #include "compat_defs.h"
 
-int HFILE_ERROR;
+WINDUMMY_FUNC(int) HFILE_ERROR;
 
 int LoadString(HINSTANCE hInstance, UINT uID, LPTSTR lpBuffer, int nBufferMax) {
     return 0;
-- 
1.6.0.4

